--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -31,8 +_,6 @@
 import net.minecraft.tags.ItemTags;
 import net.minecraft.util.Mth;
 import net.minecraft.util.SpawnUtil;
-import net.minecraft.util.profiling.Profiler;
-import net.minecraft.util.profiling.ProfilerFiller;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.InteractionHand;
@@ -216,7 +_,7 @@
             brain.setSchedule(EnvironmentAttributes.VILLAGER_ACTIVITY);
         }
 
-        brain.updateActivityFromSchedule(this.level().environmentAttributes(), this.level().getGameTime(), this.position());
+        brain.updateActivityFromSchedule(this.level().environmentAttributes(), this.level().getLevelData().getGameTime(), this.position()); // Folia - region threading - not in the world yet
     }
 
     @Override
@@ -261,10 +_,7 @@
     }
     protected void customServerAiStep(ServerLevel level, final boolean inactive) {
         // Paper end - EAR 2
-        ProfilerFiller profiler = Profiler.get();
-        profiler.push("villagerBrain");
         if (!inactive) this.getBrain().tick(level, this); // Paper - EAR 2
-        profiler.pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
         }
@@ -674,6 +_,7 @@
             this.brain.getMemory(memoryType).ifPresent(memory -> {
                 ServerLevel poiLevel = server.getLevel(memory.dimension());
                 if (poiLevel != null) {
+                    io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueOrExecuteTickTask(poiLevel, memory.pos().getX() >> 4, memory.pos().getZ() >> 4, () -> { // Folia - region threading
                     PoiManager poiManager = poiLevel.getPoiManager();
                     Optional<Holder<PoiType>> type = poiManager.getType(memory.pos());
                     BiPredicate<Villager, Holder<PoiType>> poiTypePredicate = POI_MEMORIES.get(memoryType);
@@ -681,6 +_,7 @@
                         poiManager.release(memory.pos());
                         poiLevel.debugSynchronizers().updatePoi(memory.pos());
                     }
+                    }); // Folia - region threading
                 }
             });
         }
@@ -844,7 +_,15 @@
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas start - remove streams
+        int sum = 0;
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            int i = inventory.countItem(entry.getKey()) * entry.getValue();
+            sum += i;
+        }
+        return sum;
+        // return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas end - remove streams
     }
 
     public boolean hasFarmSeeds() {
