--- a/net/minecraft/world/level/block/Portal.java
+++ b/net/minecraft/world/level/block/Portal.java
@@ -13,6 +_,10 @@
 
     @Nullable TeleportTransition getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos);
 
+    // Folia start - region threading
+    public boolean portalAsync(ServerLevel sourceWorld, Entity portalTarget, BlockPos portalPos);
+
+    // Folia end - region threading
     default Portal.Transition getLocalTransition() {
         return Portal.Transition.NONE;
     }
