--- a/net/minecraft/world/level/block/WitherSkullBlock.java
+++ b/net/minecraft/world/level/block/WitherSkullBlock.java
@@ -43,7 +_,10 @@
     }
 
     public static void checkSpawn(final Level level, final BlockPos pos) {
-        if (!level.captureBlockStates && level.getBlockEntity(pos) instanceof SkullBlockEntity placedSkull) { // CraftBukkit
+        // Folia start - region threading
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = level.getCurrentWorldData();
+        if (!worldData.captureBlockStates && level.getBlockEntity(pos) instanceof SkullBlockEntity placedSkull) { // CraftBukkit
+        // Folia end - region threading
             checkSpawn(level, pos, placedSkull);
         }
     }
