--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -118,20 +_,9 @@
 import net.minecraft.util.debug.ServerDebugSubscribers;
 import net.minecraft.util.debugchart.SampleLogger;
 import net.minecraft.util.debugchart.TpsDebugDimensions;
-import net.minecraft.util.profiling.EmptyProfileResults;
-import net.minecraft.util.profiling.ProfileResults;
-import net.minecraft.util.profiling.Profiler;
-import net.minecraft.util.profiling.ProfilerFiller;
-import net.minecraft.util.profiling.ResultField;
-import net.minecraft.util.profiling.SingleTickProfiler;
 import net.minecraft.util.profiling.jfr.Environment;
 import net.minecraft.util.profiling.jfr.JvmProfiler;
 import net.minecraft.util.profiling.jfr.callback.ProfiledDuration;
-import net.minecraft.util.profiling.metrics.profiling.ActiveMetricsRecorder;
-import net.minecraft.util.profiling.metrics.profiling.InactiveMetricsRecorder;
-import net.minecraft.util.profiling.metrics.profiling.MetricsRecorder;
-import net.minecraft.util.profiling.metrics.profiling.ServerMetricsSamplersProvider;
-import net.minecraft.util.profiling.metrics.storage.MetricsPersister;
 import net.minecraft.util.thread.ReentrantBlockableEventLoop;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.RandomSequences;
@@ -219,14 +_,7 @@
     public LevelStorageSource.LevelStorageAccess storageSource;
     protected final PlayerDataStorage playerDataStorage;
     private final SavedDataStorage savedDataStorage;
-    private final List<Runnable> tickables = Lists.newArrayList();
     // Paper - per-level GameRules
-    private MetricsRecorder metricsRecorder = InactiveMetricsRecorder.INSTANCE;
-    private Consumer<ProfileResults> onMetricsRecordingStopped = results -> this.stopRecordingMetrics();
-    private Consumer<Path> onMetricsRecordingFinished = ignored -> {};
-    private boolean willStartRecordingMetrics;
-    private MinecraftServer.@Nullable TimeProfiler debugCommandProfiler;
-    private boolean debugCommandProfilerDelayStart;
     private ServerConnectionListener connection;
     // Paper - per world load listener - moved LevelLoadListener to ServerLevel
     private @Nullable ServerStatus status;
@@ -241,8 +_,7 @@
     private volatile boolean running = true;
     public volatile boolean isRestarting = false; // Paper - flag to signify we're attempting to restart
     private boolean stopped;
-    private int tickCount;
-    private int ticksUntilAutosave = 6000;
+    // Folia - region threading
     protected final Proxy proxy;
     private boolean onlineMode;
     private boolean preventProxyConnections;
@@ -304,7 +_,7 @@
         thread.setUncaughtExceptionHandler((t, e) -> LOGGER.error("Uncaught exception in server thread", e));
         thread.setPriority(Thread.NORM_PRIORITY + 2); // Paper - Perf: Boost priority
         if (Runtime.getRuntime().availableProcessors() > 4) {
-            thread.setPriority(8);
+            thread.setPriority(10); // Canvas - bump priority of boot thread
         }
 
         S server = (S)factory.apply(thread);
@@ -318,7 +_,7 @@
     public org.bukkit.craftbukkit.CraftServer server;
     public joptsimple.OptionSet options;
     public org.bukkit.command.ConsoleCommandSender console;
-    public static int currentTick; // Paper - improve tick loop
+    // public static int currentTick; // Paper - improve tick loop // Folia - region threading
     public java.util.Queue<Runnable> processQueue = new java.util.concurrent.ConcurrentLinkedQueue<Runnable>();
     public int autosavePeriod;
     // Paper - don't store the vanilla dispatcher
@@ -335,6 +_,7 @@
     public boolean isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
     private final Set<String> pluginsBlockingSleep = new java.util.HashSet<>(); // Paper - API to allow/disallow tick sleeping
     public static final long SERVER_INIT = System.nanoTime(); // Paper - Lag compensation
+    public final dev.iyanz.aurora.engine.threadedregions.entities.EnderPearls pearls; // Canvas - region threading
     // Paper start - improve tick loop
     public final ca.spottedleaf.common.time.TickData tickTimes1s  = new ca.spottedleaf.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(1L));
     public final ca.spottedleaf.common.time.TickData tickTimes5s  = new ca.spottedleaf.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(5L));
@@ -437,46 +_,30 @@
     private static final long MAX_CHUNK_EXEC_TIME = 1000L; // 1us
     private static final long TASK_EXECUTION_FAILURE_BACKOFF = 5L * 1000L; // 5us
 
-    private long lastMidTickExecute;
-    private long lastMidTickExecuteFailure;
-
-    private boolean tickMidTickTasks() {
-        // give all worlds a fair chance at by targeting them all.
-        // if we execute too many tasks, that's fine - we have logic to correctly handle overuse of allocated time.
-        boolean executed = false;
-        for (final ServerLevel world : this.getAllLevels()) {
-            long currTime = System.nanoTime();
-            if (currTime - ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getLastMidTickFailure() <= TASK_EXECUTION_FAILURE_BACKOFF) {
-                continue;
-            }
-            if (!world.getChunkSource().pollTask()) {
-                // we need to back off if this fails
-                ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$setLastMidTickFailure(currTime);
-            } else {
-                executed = true;
-            }
-        }
-
-        return executed;
+    // Folia - region threading - moved to regionized data
+
+    private boolean tickMidTickTasks(final io.papermc.paper.threadedregions.RegionizedWorldData worldData) { // Folia - region threading
+        return worldData.world.getChunkSource().pollTask(); // Folia - region threading
     }
 
     @Override
     public final void moonrise$executeMidTickTasks() {
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData(); // Folia - region threading
         final long startTime = System.nanoTime();
-        if ((startTime - this.lastMidTickExecute) <= CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME || (startTime - this.lastMidTickExecuteFailure) <= TASK_EXECUTION_FAILURE_BACKOFF) {
+        if ((startTime - worldData.lastMidTickExecute) <= CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME || (startTime - worldData.lastMidTickExecuteFailure) <= TASK_EXECUTION_FAILURE_BACKOFF) { // Folia - region threading
             // it's shown to be bad to constantly hit the queue (chunk loads slow to a crawl), even if no tasks are executed.
             // so, backoff to prevent this
             return;
         }
 
         for (;;) {
-            final boolean moreTasks = this.tickMidTickTasks();
+            final boolean moreTasks = this.tickMidTickTasks(worldData); // Folia - region threading
             final long currTime = System.nanoTime();
             final long diff = currTime - startTime;
 
             if (!moreTasks || diff >= MAX_CHUNK_EXEC_TIME) {
                 if (!moreTasks) {
-                    this.lastMidTickExecuteFailure = currTime;
+                    worldData.lastMidTickExecuteFailure = currTime; // Folia - region threading
                 }
 
                 // note: negative values reduce the time
@@ -489,7 +_,7 @@
                 final double overuseCount = (double)overuse/(double)MAX_CHUNK_EXEC_TIME;
                 final long extraSleep = (long)Math.round(overuseCount*CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME);
 
-                this.lastMidTickExecute = currTime + extraSleep;
+                worldData.lastMidTickExecute = currTime + extraSleep; // Folia - region threading
                 return;
             }
         }
@@ -511,6 +_,49 @@
         LOGGER.warn("Performed emergency save");
     }
     // Paper end - rewrite chunk system
+    // Folia start - region threading
+    // Canvas - region threading - WTF is that?
+
+    @Override
+    public <V> CompletableFuture<V> submit(java.util.function.Supplier<V> task) {
+        if (true) {
+            throw new UnsupportedOperationException("Unsupported in region threading");
+        }
+        return super.submit(task);
+    }
+
+    @Override
+    public CompletableFuture<Void> submit(Runnable task) {
+        if (true) {
+            throw new UnsupportedOperationException("Unsupported in region threading");
+        }
+        return super.submit(task);
+    }
+
+    @Override
+    public void schedule(TickTask task) {
+        if (true) {
+            throw new UnsupportedOperationException("Unsupported in region threading");
+        }
+        super.schedule(task);
+    }
+
+    @Override
+    public void executeBlocking(Runnable runnable) {
+        if (true) {
+            throw new UnsupportedOperationException("Unsupported in region threading");
+        }
+        super.executeBlocking(runnable);
+    }
+
+    @Override
+    public void execute(Runnable runnable) {
+        if (true) {
+            throw new UnsupportedOperationException("Unsupported in region threading");
+        }
+        super.execute(runnable);
+    }
+    // Folia end - region threading
 
     public MinecraftServer(
         // CraftBukkit start
@@ -564,10 +_,11 @@
         this.tickFrame = TracyClient.createDiscontinuousFrame("Server Tick");
         this.notificationManager = notificationManager;
         this.serverActivityMonitor = new ServerActivityMonitor(notificationManager, 30);
-        this.packetProcessor = new PacketProcessor(serverThread);
+        this.packetProcessor = null; // Folia - region threading
         this.clockManager = this.getDataStorage().computeIfAbsent(ServerClockManager.TYPE);
         this.clockManager.init(this);
         this.customBossEvents = this.savedDataStorage.computeIfAbsent(CustomBossEvents.TYPE);
+        this.pearls = this.savedDataStorage.computeIfAbsent(dev.iyanz.aurora.engine.threadedregions.entities.EnderPearls.TYPE); // Canvas - region threading
         // Paper - per-level scheduledEvents
         // CraftBukkit start
         this.options = options;
@@ -833,8 +_,14 @@
             levelLoadListener.updateFocus(level.dimension(), spawnChunk);
             int height = chunkSource.getGenerator().getSpawnHeight(level);
             if (height < level.getMinY()) {
-                BlockPos worldPosition = spawnChunk.getWorldPosition();
-                height = level.getHeight(Heightmap.Types.WORLD_SURFACE, worldPosition.getX() + 8, worldPosition.getZ() + 8);
+                // Folia start - region threading
+                // pre-apply offset to make the rest easier
+                BlockPos worldPosition = spawnChunk.getWorldPosition().offset(8, 0, 8);
+                height = level.moonrise$getChunkTaskScheduler().syncLoadNonFull(
+                        worldPosition.getX() >> 4, worldPosition.getZ() >> 4,
+                        net.minecraft.world.level.chunk.status.ChunkStatus.FULL.getParent()
+                ).getHeight(Heightmap.Types.WORLD_SURFACE, worldPosition.getX(), worldPosition.getZ()) + 1; // add one to match LevelReader#getHeight
+                // Folia end - region threading
             }
 
             levelData.setSpawn(LevelData.RespawnData.of(level.dimension(), spawnChunk.getWorldPosition().offset(8, height, 8), 0.0F, 0.0F));
@@ -862,7 +_,7 @@
                 zChunkOffset += dZChunk;
             }
 
-            if (spawnBonusChest) {
+            if (false && spawnBonusChest) { // Folia - region threading - too lazy to hack around this one
                 level.registryAccess()
                     .lookup(Registries.CONFIGURED_FEATURE)
                     .flatMap(registry -> registry.get(MiscOverworldFeatures.BONUS_CHEST))
@@ -889,19 +_,19 @@
         ChunkLoadCounter chunkLoadCounter = new ChunkLoadCounter();
 
         if (true) { // CraftBukkit
-            chunkLoadCounter.track(level, () -> {
+            // chunkLoadCounter.track(level, () -> { // Folia - region threading
                 TicketStorage savedTickets = level.getDataStorage().get(TicketStorage.TYPE);
                 if (savedTickets != null) {
                     savedTickets.activateAllDeactivatedTickets();
                 }
-            });
+            // }); // Folia - region threading
         }
 
         level.levelLoadListener.start(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS, chunkLoadCounter.totalChunks()); // Paper - per world load listener
 
         do {
             level.levelLoadListener.update(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS, chunkLoadCounter.readyChunks(), chunkLoadCounter.totalChunks()); // Paper - per world load listener
-            this.executeModerately(); // CraftBukkit
+            // this.executeModerately(); // CraftBukkit // Folia - region threading
         } while (chunkLoadCounter.pendingChunks() > 0);
 
         level.levelLoadListener.finish(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS); // Paper - per world load listener
@@ -909,6 +_,7 @@
         this.updateEffectiveRespawnData();
         this.forceTicks = false; // CraftBukkit
         //level.entityManager.tick(); // SPIGOT-6526: Load pending entities so they are available to the API // Paper - rewrite chunk system
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().addWorld(level); // Folia - region threading
         new org.bukkit.event.world.WorldLoadEvent(level.getWorld()).callEvent(); // Paper - call WorldLoadEvent
     }
 
@@ -1004,7 +_,29 @@
     }
     // CraftBukkit end
 
-    protected void stopServer() {
+    // Folia start - region threading
+    public final java.util.concurrent.atomic.AtomicBoolean hasStartedShutdownThread = new java.util.concurrent.atomic.AtomicBoolean();
+
+    private void haltServerRegionThreading() {
+        if (this.hasStartedShutdownThread.getAndSet(true)) {
+            // already started shutdown
+            return;
+        }
+        new io.papermc.paper.threadedregions.RegionShutdownThread("Region shutdown thread").start();
+    }
+
+    public void stopServer() {
+        // halt scheduler
+        // don't wait, we may be on a scheduler thread
+        io.papermc.paper.threadedregions.TickRegions.getScheduler().halt(false, 0L);
+        // cannot run shutdown logic on this thread, as it may be a scheduler
+        if (true) {
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isShutdownThread()) {
+                this.haltServerRegionThreading();
+                return;
+            } // else: fall through to regular stop logic
+        }
+    // Folia end - region threading
         if (Thread.currentThread() == this.serverThread) this.executeAllRecentInternalTasks(); // Paper - execute tasks on stop
         // CraftBukkit start - prevent double stopping on multiple threads
         synchronized (this.stopLock) {
@@ -1015,10 +_,7 @@
         this.shutdownThread = Thread.currentThread(); // Paper - Improved watchdog support
         org.spigotmc.WatchdogThread.doStop(); // Paper - Improved watchdog support
         // CraftBukkit end
-        this.packetProcessor.close();
-        if (this.metricsRecorder.isRecording()) {
-            this.cancelRecordingMetrics();
-        }
+        // this.packetProcessor.close(); // Folia - region threading
 
         LOGGER.info("Stopping server");
         Commands.COMMAND_SENDING_POOL.shutdownNow(); // Paper - Perf: Async command map building; Shutdown and don't bother finishing
@@ -1029,14 +_,13 @@
             this.server.waitForAsyncTasksShutdown(); // Paper - Wait for Async Tasks during shutdown
         }
         // CraftBukkit end
-        this.getConnection().stop();
-        this.isSaving = true;
-        if (this.playerList != null) {
-            LOGGER.info("Saving players");
-            this.playerList.saveAll();
-            this.playerList.removeAll(this.isRestarting); // Paper
-            this.getConnection().handleAllDisconnections(); // Paper
+        this.isSaving = true; // Canvas - region threading
+        // Folia start - region threading
+        if (true) {
+            // the rest till part 2 is handled by the region shutdown thread
+            return;
         }
+        // Folia end - region threading
 
         LOGGER.info("Saving worlds");
 
@@ -1067,6 +_,11 @@
         this.saveAllChunks(false, true, false, true); // Paper - rewrite chunk system
 
         this.isSaving = false;
+    // Folia start - region threading
+        this.stopPart2();
+    }
+    public void stopPart2() {
+    // Folia end - region threading
         this.savedDataStorage.close();
         this.resources.close();
 
@@ -1121,6 +_,7 @@
         if (this.isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread("Server stopped"); // Paper - Debugging
         // Paper end
         this.running = false;
+        this.stopServer(); // Folia - region threading
         if (wait) {
             try {
                 this.serverThread.join();
@@ -1173,10 +_,6 @@
 
         // note: To avoid possibly spinning forever, only execute tasks that are roughly available at the beginning
         //       of this call. Packet processing and chunk system tasks are possibly always being queued.
-        final ProfilerFiller profiler = Profiler.get();
-        profiler.push("moonrise:run_all_tasks");
-
-        profiler.push("moonrise:run_all_server");
         // avoid calling pollTask - we just want to execute queued tasks
         final int currentTick = this.tickTaskTickCount.incrementAndGet();
         final java.util.function.Predicate<net.minecraft.server.TickTask> taskPredicate = (final TickTask task) -> {
@@ -1190,38 +_,23 @@
             ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this).moonrise$executeMidTickTasks(); // Paper - rewrite chunk system
             this.packetProcessor.executeSinglePacket();
         }
-        profiler.popPush("moonrise:run_all_packets");
         while (this.packetProcessor.executeSinglePacket()) {
             // execute possibly latency sensitive chunk system tasks (see above)
             ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this).moonrise$executeMidTickTasks(); // Paper - rewrite chunk system
         }
-        profiler.popPush("moonrise:run_all_chunk");
         // Paper start - rewrite chunk system
         for (final ServerLevel world : this.getAllLevels()) {
-            profiler.push(world.toString() + " " + world.dimension().identifier()); // keep same formatting from regular tick, see tickChildren
-
             // note: legacy tasks may expect a distance manager update
-            profiler.push("moonrise:distance_manager_update");
             ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getChunkTaskScheduler().chunkHolderManager.processTicketUpdates();
-            profiler.popPush("moonrise:legacy_chunk_tasks");
             world.getChunkSource().mainThreadProcessor.executeAllRecentInternalTasks();
-            profiler.popPush("moonrise:chunk_system_tasks");
             ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getChunkTaskScheduler().executeAllRecentlyQueuedMainThreadTasks();
-            profiler.pop();
-
-            profiler.pop(); // world name
         }
         // Paper end - rewrite chunk system
-        profiler.pop(); // moonrise:run_all_chunk
-        profiler.pop(); // moonrise:run_all_tasks
 
         this.finishMeasuringTaskExecutionTime();
     }
 
     private void recordTaskExecutionTimeWhileWaiting() {
-        final ProfilerFiller profiler = Profiler.get();
-
-        profiler.push("moonrise:execute_tasks_until_tick");
         this.waitingForNextTick = true;
         // implement waitForTasks
         final boolean isLoggingEnabled = this.isTickTimeLoggingEnabled();
@@ -1257,7 +_,6 @@
         } finally {
             this.waitingForNextTick = false;
         }
-        profiler.pop();
     }
     // Paper end - improve tick loop
 
@@ -1273,6 +_,14 @@
             this.status = this.buildServerStatus();
 
             this.server.spark.enableBeforePlugins(); // Paper - spark
+            // Folia start - region threading
+            if (true) {
+                io.papermc.paper.threadedregions.RegionizedServer.getInstance().init(); // Folia - region threading - only after loading worlds
+                final long actualDoneTimeMs = System.currentTimeMillis() - org.bukkit.craftbukkit.Main.BOOT_TIME.toEpochMilli(); // Paper - Improve startup message
+                LOGGER.info("Done ({})! For help, type \"help\"", String.format(java.util.Locale.ROOT, "%.3fs", actualDoneTimeMs / 1000.00D)); // Paper - Improve startup message
+                return; // Canvas - region threading
+            }
+            // Folia end - region threading
             // Spigot start
             // Paper start
             LOGGER.info("Running delayed init tasks");
@@ -1296,11 +_,7 @@
                 LOGGER.info("*************************************************************************************");
             }
             // Paper end - Add onboarding message for initial server start
-            // Paper start - Improve outdated version checking
-            if (System.getProperty("paper.disableStartupVersionCheck") == null && io.papermc.paper.configuration.GlobalConfiguration.get().updateChecker.enabled) {
-                CompletableFuture.runAsync(com.destroystokyo.paper.PaperVersionFetcher::getUpdateStatusStartupMessage, io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR);
-            }
-            // Paper end - Improve outdated version checking
+            // Canvas - Rebrand
 
             while (this.running) {
                 final long tickStart = System.nanoTime(); // Paper - improve tick loop
@@ -1333,21 +_,15 @@
                 this.lastOverloadWarningNanos = this.nextTickTimeNanos;
 
                 this.currentTickStart = tickStart;
-                ++MinecraftServer.currentTick;
+                // ++MinecraftServer.currentTick; // Folia - region threading
                 // Paper end - improve tick loop
 
                 boolean sprinting = thisTickNanos == 0L;
-                if (this.debugCommandProfilerDelayStart) {
-                    this.debugCommandProfilerDelayStart = false;
-                    this.debugCommandProfiler = new MinecraftServer.TimeProfiler(Util.getNanos(), this.tickCount);
-                }
 
                 // Paper - improve tick loop - done above
 
-                try (Profiler.Scope ignored = Profiler.use(this.createProfiler())) {
+                {
                     this.processPacketsAndTick(sprinting);
-                    ProfilerFiller profiler = Profiler.get();
-                    profiler.push("nextTickWait");
                     this.mayHaveDelayedTasks = true;
                     this.delayedTasksMaxNextTickTimeNanos = Math.max(Util.getNanos() + thisTickNanos, this.nextTickTimeNanos);
                     this.startMeasuringTaskExecutionTime();
@@ -1357,10 +_,7 @@
                         this.tickRateManager.endTickWork();
                     }
 
-                    profiler.pop();
                     this.logFullTickTime();
-                } finally {
-                    this.endMetricsRecordingTick();
                 }
 
                 this.isReady = true;
@@ -1378,7 +_,7 @@
             }
 
             this.onServerCrash(report);
-        } finally {
+        } if (false) { // Canvas - region threading
             try {
                 this.stopped = true;
                 this.stopServer();
@@ -1476,41 +_,21 @@
     @Override
     // Paper start - anything that does try to post to main during watchdog crash, run on watchdog
     public TickTask wrapRunnable(Runnable runnable) {
-        if (this.hasStopped && Thread.currentThread().equals(this.shutdownThread)) {
-            runnable.run();
-            runnable = () -> {};
-        }
-        // Paper end - anything that does try to post to main during watchdog crash, run on watchdog
-        return new TickTask(this.tickTaskTickCount.get(), runnable); // Paper - use different tick field for tick tasks (see #shouldRun)
+        throw new UnsupportedOperationException("Unsupported in region threading"); // Folia - region threading
     }
 
     @Override
     protected boolean shouldRun(final TickTask task) {
-        // Paper start - improve tick loop - do not stall queued tasks
-        // note: make this overflow safe as well
-        return this.tickTaskTickCount.getPlain() - task.getTick() > 0 ||
-            /*
-             * Ensure that we execute any task as long as we are waiting for the next tick.
-             * The Vanilla server will use managedBlock when awaiting the next tick, but
-             * we do not. The Vanilla managedBlock function will bypass task execution
-             * checks, and in order to ensure we execute tasks like Vanilla we need to also
-             * bypass task execution checks.
-             * This fixes {@link #recordTaskExecutionTimeWhileWaiting} not executing tasks
-             * that it should be executing.
-             */
-            this.waitingForNextTick ||
-            this.haveTime();
-        // Paper end - improve tick loop - do not stall queued tasks
+        throw new UnsupportedOperationException("Unsupported in region threading"); // Folia - region threading
     }
 
     @Override
     protected boolean pollTask() {
-        boolean mayHaveMoreTasks = this.packetProcessor.executeSinglePacket() | this.pollTaskInternal(); // Paper - improve tick loop - process packets while waiting inbetween ticks
-        this.mayHaveDelayedTasks = mayHaveMoreTasks;
-        return mayHaveMoreTasks;
+        throw new UnsupportedOperationException("Unsupported in region threading"); // Folia - region threading
     }
 
     private boolean pollTaskInternal() {
+        if (true) throw new UnsupportedOperationException("Unsupported in region threading"); // Folia - region threading
         if (super.pollTask()) {
             this.moonrise$executeMidTickTasks(); // Paper - rewrite chunk system
             return true;
@@ -1530,8 +_,7 @@
 
     @Override
     protected void doRunTask(final TickTask task) {
-        Profiler.get().incrementCounter("runTask");
-        super.doRunTask(task);
+        throw new UnsupportedOperationException("Unsupported in region threading"); // Folia - region threading
     }
 
     private Optional<ServerStatus.Favicon> loadStatusIcon() {
@@ -1576,12 +_,15 @@
         return false;
     }
 
-    protected void tickServer(final BooleanSupplier haveTime) {
+    // Folia start - region threading
+    public void tickServer(final long startTime, final long scheduledEnd, final long targetBuffer,
+                           final io.papermc.paper.threadedregions.TickRegions.TickRegionData region) {
+    // Folia end - region threading
         org.spigotmc.WatchdogThread.tick(); // Spigot
-        long nano = Util.getNanos();
+        long nano = startTime; // Folia - region threading
         int emptyTickThreshold = this.pauseWhenEmptySeconds() * 20;
         this.removeDisabledPluginsBlockingSleep(); // Paper - API to allow/disallow tick sleeping
-        if (emptyTickThreshold > 0) {
+        if (false && emptyTickThreshold > 0) { // Folia - region threading - this is complicated to implement, and even if done correctly is messy
             if (this.playerList.getPlayerCount() == 0 && !this.tickRateManager.isSprinting() && this.pluginsBlockingSleep.isEmpty()) { // Paper - API to allow/disallow tick sleeping
                 this.emptyTicks++;
             } else {
@@ -1606,76 +_,115 @@
                     level.getChunkSource().tick(() -> true, false);
                 }
                 // Paper end - avoid issues with certain tasks not processing during sleep
-                this.server.spark.executeMainThreadTasks(); // Paper - spark
+                // this.server.spark.executeMainThreadTasks(); // Paper - spark // Folia - region threading
                 this.tickConnection();
                 this.server.spark.tickEnd(((double)(System.nanoTime() - this.currentTickStart) / 1000000D)); // Paper - spark
                 return;
             }
         }
 
+        // Folia start - region threading
+        region.world.getCurrentWorldData().updateTickData();
+        BooleanSupplier haveTime = () -> {
+            return scheduledEnd - System.nanoTime() > targetBuffer;
+        };
+
+        // Folia end - region threading
         this.server.spark.tickStart(); // Paper - spark
-        new com.destroystokyo.paper.event.server.ServerTickStartEvent(this.tickCount+1).callEvent(); // Paper - Server Tick Events
-        this.tickCount++;
-        this.tickRateManager.tick();
-        this.tickChildren(haveTime);
-        if (nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
+        // Folia start - region threading
+        new com.destroystokyo.paper.event.server.ServerTickStartEvent((int)region.getCurrentTick()).callEvent(); // Paper - Server Tick Events
+        if (region != null) {
+            region.getTaskQueueData().drainTasks();
+            // run all player packets
+            for (final ServerPlayer player : new java.util.ArrayList<>(region.world.getCurrentWorldData().getLocalPlayers())) {
+                PacketProcessor packetProcessor = player.getBukkitEntity().packetProcessor;
+                while (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) && packetProcessor.executeSinglePacket());
+            }
+            // Canvas start - region threading
+            // we process end credits players HERE
+            for (final ServerPlayer player : new java.util.ArrayList<>(region.world.getCurrentWorldData().endCreditsPlayers)) {
+                PacketProcessor packetProcessor = player.getBukkitEntity().packetProcessor;
+                while (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) && packetProcessor.executeSinglePacket());
+            }
+            // Canvas end - region threading
+            ((io.papermc.paper.threadedregions.scheduler.FoliaRegionScheduler)org.bukkit.Bukkit.getRegionScheduler()).tick();
+            // now run all the entity schedulers
+            for (io.papermc.paper.threadedregions.EntityScheduler scheduler : region.world.getCurrentWorldData().entitySchedulerTickList.getAllSchedulers()) {
+                net.minecraft.world.entity.Entity handle = scheduler.entity.getHandleRaw();
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(handle) || scheduler.isRetired()) {
+                    continue;
+                }
+
+                scheduler.executeTick();
+            }
+            // fix passenger desync
+            for (final net.minecraft.world.entity.Entity entity : region.world.getCurrentWorldData().getLocalEntitiesCopy()) {
+                entity.fixPassengerDesync();
+            }
+        }
+        // Canvas start - region threading
+        if (region.world.canvas$unloadTicket.isPresent()) {
+            region.getRegionSchedulingHandle().markNonSchedulable();
+            return;
+        }
+        // Canvas end - region threading
+        // Folia end - region threading
+        //this.tickCount++; // Folia - region threading
+        //this.tickRateManager.tick(); // Folia - region threading
+        // Canvas start - region threading
+        try {
+            this.tickChildren(haveTime, region);
+        } catch (final Throwable thrown) {
+            CrashReport crashReport = CrashReport.forThrowable(thrown, "Exception ticking region");
+            region.world.fillReportDetails(crashReport);
+            throw new ReportedException(crashReport);
+        }
+        // Canvas end - region threading
+        if (false && nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) { // Folia - region threading
             this.lastServerStatus = nano;
             this.status = this.buildServerStatus();
         }
 
-        this.ticksUntilAutosave--;
+        // this.ticksUntilAutosave--; // Folia - region threading
         // Paper start - Incremental chunk and player saving
         int playerSaveInterval = io.papermc.paper.configuration.GlobalConfiguration.get().playerAutoSave.rate;
         if (playerSaveInterval < 0) {
             playerSaveInterval = this.autosavePeriod;
         }
-        Profiler.get().push("save");
-        final boolean fullSave = this.autosavePeriod > 0 && this.tickCount % this.autosavePeriod == 0;
+        final boolean fullSave = this.autosavePeriod > 0 && io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() % this.autosavePeriod == 0; // Folia - region threading
         try {
             this.isSaving = true;
-            if (playerSaveInterval > 0) {
+            if (dev.iyanz.aurora.engine.GlobalConfiguration.getInstance().autosave.autosavePlayers && playerSaveInterval > 0) { // Canvas - region threading
                 this.playerList.saveAll(playerSaveInterval);
             }
-            if (fullSave) {
+            if (false && fullSave) { // Folia - region threading - don't auto save level data
                 this.saveGlobalData(false);
             }
-            for (final ServerLevel level : this.getAllLevels()) {
+            for (final ServerLevel level : Arrays.asList(region.world)) { // Folia - region threading
                 if (level.paperConfig().chunks.autoSaveInterval.value() > 0) {
-                    level.saveIncrementally(fullSave);
+                    level.saveIncrementally(false); // Folia - region threading - don't auto save level data
                 }
             }
         } finally {
             this.isSaving = false;
         }
-        Profiler.get().pop();
         // Paper end - Incremental chunk and player saving
 
-        ProfilerFiller profiler = Profiler.get();
-        this.server.spark.executeMainThreadTasks(); // Paper - spark
+        // this.server.spark.executeMainThreadTasks(); // Paper - spark // Folia - region threading
         // Paper start - Server Tick Events
         long endTime = System.nanoTime();
-        long remaining = this.nextTickTimeNanos - endTime;
-        new com.destroystokyo.paper.event.server.ServerTickEndEvent(this.tickCount, ((double)(endTime - this.currentTickStart) / 1000000D), remaining).callEvent();
+        long remaining = scheduledEnd - endTime; // Folia - region threading
+        new com.destroystokyo.paper.event.server.ServerTickEndEvent((int)io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick(), ((double)(endTime - startTime) / 1000000D), remaining).callEvent(); // Folia - region threading
         // Paper end - Server Tick Events
-        this.server.spark.tickEnd(((double)(endTime - this.currentTickStart) / 1000000D)); // Paper - spark
-        profiler.push("tallying");
-        long tickTime = Util.getNanos() - nano;
-        int tickIndex = this.tickCount % 100;
-        this.aggregatedTickTimesNanos = this.aggregatedTickTimesNanos - this.tickTimesNanos[tickIndex];
-        this.aggregatedTickTimesNanos += tickTime;
-        this.tickTimesNanos[tickIndex] = tickTime;
-        this.smoothedTickTimeMillis = this.smoothedTickTimeMillis * 0.8F + (float)tickTime / (float)TimeUtil.NANOSECONDS_PER_MILLISECOND * 0.19999999F;
-        this.logTickMethodTime(nano);
-        profiler.pop();
+        this.server.spark.tickEnd(((double)(endTime - startTime) / 1000000D)); // Paper - spark // Folia - region threading
+        // Folia - region threading
     }
 
     protected void processPacketsAndTick(final boolean sprinting) {
-        ProfilerFiller profiler = Profiler.get();
-        profiler.push("tick");
         this.tickFrame.start();
         // Paper - improve tick loop - moved into runAllTasksAtTickStart
         this.runAllTasksAtTickStart(); // Paper - improve tick loop
-        this.tickServer(sprinting ? () -> false : this::haveTime);
+        if (true) throw new UnsupportedOperationException("Unsupported in region threading"); // Folia - region threading
         // Paper start - rewrite chunk system
         final Throwable crash = this.chunkSystemCrash;
         if (crash != null) {
@@ -1685,16 +_,12 @@
         // Paper end - rewrite chunk system
         this.tickFrame.end();
         this.recordEndOfTick(); // Paper - improve tick loop
-        profiler.pop();
     }
 
     private void autoSave() {
-        this.ticksUntilAutosave = this.autosavePeriod; // CraftBukkit
+        // this.ticksUntilAutosave = this.autosavePeriod; // CraftBukkit // Folia - region threading
         LOGGER.debug("Autosave started");
-        ProfilerFiller profiler = Profiler.get();
-        profiler.push("save");
         this.saveEverything(true, false, false);
-        profiler.pop();
         LOGGER.debug("Autosave finished");
     }
 
@@ -1704,30 +_,22 @@
         }
     }
 
-    private int computeNextAutosaveInterval() {
-        float ticksPerSecond;
-        if (this.tickRateManager.isSprinting()) {
-            long estimatedTickTimeNanos = this.getAverageTickTimeNanos() + 1L;
-            ticksPerSecond = (float)TimeUtil.NANOSECONDS_PER_SECOND / (float)estimatedTickTimeNanos;
-        } else {
-            ticksPerSecond = this.tickRateManager.tickrate();
-        }
-
-        int intendedIntervalInSeconds = 300;
-        return Math.max(100, (int)(ticksPerSecond * 300.0F));
-    }
+    // Folia - region threading - use absolute time instead of this
 
     public void onTickRateChanged() {
-        int newAutosaveInterval = this.computeNextAutosaveInterval();
-        if (newAutosaveInterval < this.ticksUntilAutosave) {
-            this.ticksUntilAutosave = newAutosaveInterval;
-        }
+        // Folia - region threading - use absolute time instead of this
     }
 
     protected abstract SampleLogger getTickTimeLogger();
 
     public abstract boolean isTickTimeLoggingEnabled();
 
+    // Folia start - region threading
+    public void rebuildServerStatus() {
+        this.status = this.buildServerStatus();
+    }
+
+    // Folia end - region threading
     private ServerStatus buildServerStatus() {
         ServerStatus.Players players = this.buildPlayerStatus();
         return new ServerStatus(
@@ -1740,7 +_,7 @@
     }
 
     private ServerStatus.Players buildPlayerStatus() {
-        List<ServerPlayer> players = this.playerList.getPlayers();
+        List<ServerPlayer> players = new java.util.ArrayList<>(this.playerList.getPlayers()); // Folia - region threading
         int maxPlayers = this.getMaxPlayers();
         if (this.hidesOnlinePlayers()) {
             return new ServerStatus.Players(maxPlayers, players.size(), List.of());
@@ -1759,14 +_,21 @@
         return new ServerStatus.Players(maxPlayers, players.size(), sample);
     }
 
-    public final io.papermc.paper.threadedregions.EntityScheduler.EntitySchedulerTickList entitySchedulerTickList = new io.papermc.paper.threadedregions.EntityScheduler.EntitySchedulerTickList(); // Paper - optimise Folia entity scheduler
+    // Folia - region threading - moved to regionized data
 
-    protected void tickChildren(final BooleanSupplier haveTime) {
-        ProfilerFiller profiler = Profiler.get();
-        this.getPlayerList().getPlayers().forEach(playerx -> playerx.connection.suspendFlushing());
-        this.server.getScheduler().mainThreadHeartbeat(); // CraftBukkit
+    // Folia start - region threading
+    protected void tickChildren(final BooleanSupplier haveTime, final io.papermc.paper.threadedregions.TickRegions.TickRegionData region) {
+        final io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        // this.getPlayerList().getPlayers().forEach(playerx -> playerx.connection.suspendFlushing());
+        // this.server.getScheduler().mainThreadHeartbeat(); // CraftBukkit
+        final List<ServerPlayer> suspendedFlushes = new java.util.ArrayList<>(regionizedWorldData.getLocalPlayers());
+        for (ServerPlayer player : suspendedFlushes) {
+            player.connection.suspendFlushing();
+        }
+    // Folia end - region threading
         // Paper start - optimise Folia entity scheduler
-        ((io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler)org.bukkit.Bukkit.getGlobalRegionScheduler()).tick();
+        //((io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler)org.bukkit.Bukkit.getGlobalRegionScheduler()).tick(); // Folia - region threading - moved to global tick
+        /* // Folia start - region threading - move to tickRegion
         for (io.papermc.paper.threadedregions.EntityScheduler scheduler : this.entitySchedulerTickList.getAllSchedulers()) {
             if (scheduler.isRetired()) {
                 continue;
@@ -1775,14 +_,12 @@
             scheduler.executeTick();
         }
         // Paper end - optimise Folia entity scheduler
+        */ // Folia end - region threading - move to tickRegion
         // Paper end - Folia scheduler API
-        io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.ADVENTURE_CLICK_MANAGER.handleQueue(this.tickCount); // Paper
-        io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.DIALOG_CLICK_MANAGER.handleQueue(this.tickCount); // Paper
-        profiler.push("commandFunctions");
-        this.getFunctions().tick();
-        profiler.pop();
-        if (this.tickRateManager.runsNormally()) {
-            profiler.push("clocks");
+        // io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.ADVENTURE_CLICK_MANAGER.handleQueue(this.tickCount); // Paper // Folia - region threading - moved to global tick
+        // io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.DIALOG_CLICK_MANAGER.handleQueue(this.tickCount); // Paper // Folia - region threading - moved to global tick
+        // this.getFunctions().tick(); // Folia - region threading - TODO Purge functions
+        if (false && this.tickRateManager.runsNormally()) { // Folia - region threading - move to global tick
             // Paper start - per-world time
             if (io.papermc.paper.configuration.GlobalConfiguration.get().time.affectsAllWorlds) {
                 this.clockManager.tick();
@@ -1792,72 +_,82 @@
                 }
             }
             // Paper end - per-world time
-            profiler.pop();
         }
 
-        if (this.tickCount % 20 == 0) {
-            profiler.push("timeSync");
+        if (true) { // Folia - region threading
             this.forceGameTimeSynchronization();
-            profiler.pop();
         }
 
         // CraftBukkit start
         // Run tasks that are waiting on processing
-        while (!this.processQueue.isEmpty()) {
+        if (false) while (!this.processQueue.isEmpty()) { // Folia - region threading
             this.processQueue.remove().run();
         }
+        region.world.getCurrentWorldData().getCanvasWorldData().tpsbar.tick(); // Canvas - tpsbar
+        region.world.getCurrentWorldData().getCanvasWorldData().rambar.tick(); // Canvas - rambar
 
-        profiler.push("levels");
         this.updateEffectiveRespawnData();
 
-        this.isIteratingOverLevels = true; // Paper - Throw exception on world create while being ticked
-        for (ServerLevel level : this.getAllLevels()) {
-            level.hasPhysicsEvent = org.bukkit.event.block.BlockPhysicsEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - BlockPhysicsEvent
-            level.hasEntityMoveEvent = io.papermc.paper.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - Add EntityMoveEvent
-            level.updateLagCompensationTick(); // Paper - lag compensation
-            net.minecraft.world.level.block.entity.HopperBlockEntity.skipHopperEvents = level.paperConfig().hopper.disableMoveEvent || org.bukkit.event.inventory.InventoryMoveItemEvent.getHandlerList().getRegisteredListeners().length == 0; // Paper - Perf: Optimize Hoppers
-            profiler.push(() -> level + " " + level.dimension().identifier());
-            profiler.push("tick");
-
-            try {
-                level.tick(haveTime);
+        // Canvas start - region threading - save all command
+        region.canvas$saveAllTicket.consumeIfPresent(ticket -> {
+            try {
+                ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)region.world).moonrise$getChunkTaskScheduler().chunkHolderManager.saveAllChunksRegionised(
+                    ticket.flush(), false, false, false, false, true, false
+                );
+                this.getPlayerList().saveAll();
+            } catch (final Throwable thrown) {
+                ticket.exceptionPropagator().accept(thrown);
+            } finally {
+                ticket.callback().run();
+            }
+        });
+
+        // Canvas end - region threading - save all command
+        // Folia start - region threading
+        // this.isIteratingOverLevels = true; // Paper - Throw exception on world create while being ticked
+        if (true) { ServerLevel level = region.world;
+            // level.hasPhysicsEvent = org.bukkit.event.block.BlockPhysicsEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - BlockPhysicsEvent
+            // level.hasEntityMoveEvent = io.papermc.paper.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - Add EntityMoveEvent
+            // level.updateLagCompensationTick(); // Paper - lag compensation
+            // net.minecraft.world.level.block.entity.HopperBlockEntity.skipHopperEvents = level.paperConfig().hopper.disableMoveEvent || org.bukkit.event.inventory.InventoryMoveItemEvent.getHandlerList().getRegisteredListeners().length == 0; // Paper - Perf: Optimize Hoppers
+        // Folia end - region threading
+
+            try {
+                level.tick(haveTime, region); // Folia - region threading
             } catch (Throwable t) {
                 CrashReport report = CrashReport.forThrowable(t, "Exception ticking world");
                 level.fillReportDetails(report);
                 throw new ReportedException(report);
             }
 
-            profiler.pop();
-            profiler.pop();
-            level.explosionDensityCache.clear(); // Paper - Optimize explosions
+            regionizedWorldData.explosionDensityCache.clear(); // Paper - Optimize explosions // Folia - region threading
         }
-        this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
+        // this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked // Folia - region threading
 
-        profiler.popPush("connection");
-        this.tickConnection();
-        profiler.popPush("players");
-        this.playerList.tick();
-        profiler.popPush("debugSubscribers");
-        this.debugSubscribers.tick();
+        // Folia & Canvas start - region threading
+        regionizedWorldData.tickConnections();
+        // this.playerList.tick(); // moved to global tick
+        // debug subscribers and the game test framework are not safe for region threading
+        /* this.debugSubscribers.tick();
         if (this.tickRateManager.runsNormally()) {
-            profiler.popPush("gameTests");
             GameTestTicker.SINGLETON.tick();
-        }
-
-        profiler.popPush("server gui refresh");
-
-        for (Runnable tickable : this.tickables) {
+        } */
+
+        // tickables are blatantly unsafe... what the fuck is this thing
+        /* for (Runnable tickable : this.tickables) {
             tickable.run();
-        }
-
-        profiler.popPush("send chunks");
-
-        for (ServerPlayer player : this.playerList.getPlayers()) {
-            player.connection.chunkSender.sendNextChunks(player);
+        } */
+        // Folia & Canvas end - region threading
+
+        // Folia start - region threading
+        for (final ServerPlayer player : suspendedFlushes) {
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player)) {
+                continue;
+            }
+        // Folia end - region threading
             player.connection.resumeFlushing();
         }
 
-        profiler.pop();
         this.serverActivityMonitor.tick();
     }
 
@@ -1875,18 +_,19 @@
     }
 
     public void forceGameTimeSynchronization() {
-        ProfilerFiller profiler = Profiler.get();
-        profiler.push("timeSync");
-        // Paper start - per-world time
-        for (ServerLevel level : this.getAllLevels()) {
-            this.playerList.broadcastAll(new ClientboundSetTimePacket(level.getGameTime(), Map.of()), level);
+        // Folia start - region threading
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        ClientboundSetTimePacket timePacket = new ClientboundSetTimePacket(worldData.world.getGameTime(), Map.of());
+        for (ServerPlayer player : worldData.getLocalPlayers()) {
+            if ((io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() + player.getId()) % 20 == 0) {
+                player.connection.send(timePacket);
+            }
         }
-        // Paper end - per-world time
-        profiler.pop();
+        // Folia end - region threading
     }
 
     public void addTickable(final Runnable tickable) {
-        this.tickables.add(tickable);
+        throw new UnsupportedOperationException("Unsupported in region threading"); // Canvas - region threading
     }
 
     protected void setId(final String serverId) {
@@ -1948,7 +_,7 @@
     }
 
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return dev.iyanz.aurora.engine.GlobalConfiguration.getInstance().serverModName; // Canvas - Use configurable server branding
     }
 
     public ServerClockManager clockManager() {
@@ -2158,11 +_,13 @@
         int count = 0;
 
         for (ServerPlayer player : this.getPlayerList().getPlayers()) {
+            player.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer entityPlayer) -> { // Folia & Canvas - region threading
             // Paper start - Expand PlayerGameModeChangeEvent
-            org.bukkit.event.player.PlayerGameModeChangeEvent event = player.setGameMode(gameType, org.bukkit.event.player.PlayerGameModeChangeEvent.Cause.DEFAULT_GAMEMODE, null);
+            org.bukkit.event.player.PlayerGameModeChangeEvent event = entityPlayer.setGameMode(gameType, org.bukkit.event.player.PlayerGameModeChangeEvent.Cause.DEFAULT_GAMEMODE, null); // Folia - region threading
             if (event == null || event.isCancelled()) {
-                continue;
+                return; // Folia - region threading
             }
+            }); // Folia & Canvas - region threading
             count++;
             // Paper end - Expand PlayerGameModeChangeEvent
         }
@@ -2187,7 +_,7 @@
     }
 
     public int getTickCount() {
-        return this.tickCount;
+        throw new UnsupportedOperationException("Unsupported in region threading"); // Folia - region threading
     }
 
     public boolean isUnderSpawnProtection(final ServerLevel level, final BlockPos pos, final Player player) {
@@ -2223,6 +_,15 @@
     }
 
     public void invalidateStatus() {
+        // Folia start - region threading
+        if (true) {
+            // we don't need this to notify the global tick region
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTaskWithoutNotify(() -> {
+                io.papermc.paper.threadedregions.RegionizedServer.getInstance().invalidateStatus();
+            });
+            return;
+        }
+        // Folia end - region threading
         this.lastServerStatus = 0L;
     }
 
@@ -2237,6 +_,7 @@
 
     @Override
     public void executeIfPossible(final Runnable command) {
+        if (true) throw new UnsupportedOperationException("Unsupported in region threading"); // Folia - region threading
         if (this.isStopped()) {
             throw new io.papermc.paper.util.ServerStopRejectedExecutionException("Server already shutting down"); // Paper - do not prematurely disconnect players on stop
         }
@@ -2253,6 +_,13 @@
         return 256;
     }
 
+    // Canvas start - no chat reports
+    public boolean canvas$ncrEnforceSecureProfile() {
+        if (dev.iyanz.aurora.engine.GlobalConfiguration.getInstance().chat.disableChatReporting) return false; // NCR doesn't require secure profile
+        return this.enforceSecureProfile();
+    }
+
+    // Canvas end - no chat reports
     public boolean enforceSecureProfile() {
         return false;
     }
@@ -2569,6 +_,7 @@
 
     // Paper start - per-level WeatherData
     public void setWeatherParameters(final ServerLevel level, final int clearTime, final int rainTime, final boolean raining, final boolean thundering) {
+        io.papermc.paper.threadedregions.RegionizedServer.ensureGlobalTickThread("Cannot set weather params async"); // Canvas - region threading
         WeatherData weatherData = level.getWeatherData();
         // Paper end - per-level WeatherData
         weatherData.setClearWeatherTime(clearTime);
@@ -2608,7 +_,12 @@
     }
 
     public long getAverageTickTimeNanos() {
-        return this.aggregatedTickTimesNanos / Math.min(100, Math.max(this.tickCount, 1));
+        // Folia start - region threading
+        if (io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentTickingTask() instanceof io.papermc.paper.threadedregions.TickRegionScheduler.RegionScheduleHandle handle) {
+            return (long)Math.ceil(handle.getTickReport5s(System.nanoTime()).timePerTickData().segmentAll().average());
+        }
+        return 0L;
+        // Folia end - region threading
     }
 
     public long[] getTickTimesNanos() {
@@ -2745,55 +_,6 @@
     }
     // CraftBukkit end
 
-    private ProfilerFiller createProfiler() {
-        if (this.willStartRecordingMetrics) {
-            this.metricsRecorder = ActiveMetricsRecorder.createStarted(
-                new ServerMetricsSamplersProvider(Util.timeSource(), this.isDedicatedServer()),
-                Util.timeSource(),
-                Util.ioPool(),
-                new MetricsPersister("server"),
-                this.onMetricsRecordingStopped,
-                reportPath -> {
-                    this.executeBlocking(() -> this.saveDebugReport(reportPath.resolve("server")));
-                    this.onMetricsRecordingFinished.accept(reportPath);
-                }
-            );
-            this.willStartRecordingMetrics = false;
-        }
-
-        this.metricsRecorder.startTick();
-        return SingleTickProfiler.decorateFiller(this.metricsRecorder.getProfiler(), SingleTickProfiler.createTickProfiler("Server"));
-    }
-
-    protected void endMetricsRecordingTick() {
-        this.metricsRecorder.endTick();
-    }
-
-    public boolean isRecordingMetrics() {
-        return this.metricsRecorder.isRecording();
-    }
-
-    public void startRecordingMetrics(final Consumer<ProfileResults> onStopped, final Consumer<Path> onFinished) {
-        this.onMetricsRecordingStopped = report -> {
-            this.stopRecordingMetrics();
-            onStopped.accept(report);
-        };
-        this.onMetricsRecordingFinished = onFinished;
-        this.willStartRecordingMetrics = true;
-    }
-
-    public void stopRecordingMetrics() {
-        this.metricsRecorder = InactiveMetricsRecorder.INSTANCE;
-    }
-
-    public void finishRecordingMetrics() {
-        this.metricsRecorder.end();
-    }
-
-    public void cancelRecordingMetrics() {
-        this.metricsRecorder.cancel();
-    }
-
     public Path getWorldPath(final LevelResource resource) {
         return this.storageSource.getLevelPath(resource);
     }
@@ -2842,24 +_,6 @@
         return this.isSaving;
     }
 
-    public boolean isTimeProfilerRunning() {
-        return this.debugCommandProfilerDelayStart || this.debugCommandProfiler != null;
-    }
-
-    public void startTimeProfiler() {
-        this.debugCommandProfilerDelayStart = true;
-    }
-
-    public ProfileResults stopTimeProfiler() {
-        if (this.debugCommandProfiler == null) {
-            return EmptyProfileResults.EMPTY;
-        }
-
-        ProfileResults results = this.debugCommandProfiler.stop(Util.getNanos(), this.tickCount);
-        this.debugCommandProfiler = null;
-        return results;
-    }
-
     public int getMaxChainedNeighborUpdates() {
         return 1000000;
     }
@@ -3087,55 +_,6 @@
     public record ServerResourcePackInfo(UUID id, String url, String hash, boolean isRequired, @Nullable Component prompt) {
     }
 
-    private static class TimeProfiler {
-        private final long startNanos;
-        private final int startTick;
-
-        private TimeProfiler(final long startNanos, final int startTick) {
-            this.startNanos = startNanos;
-            this.startTick = startTick;
-        }
-
-        private ProfileResults stop(final long stopNanos, final int stopTick) {
-            return new ProfileResults() {
-                @Override
-                public List<ResultField> getTimes(final String path) {
-                    return Collections.emptyList();
-                }
-
-                @Override
-                public boolean saveResults(final Path file) {
-                    return false;
-                }
-
-                @Override
-                public long getStartTimeNano() {
-                    return TimeProfiler.this.startNanos;
-                }
-
-                @Override
-                public int getStartTimeTicks() {
-                    return TimeProfiler.this.startTick;
-                }
-
-                @Override
-                public long getEndTimeNano() {
-                    return stopNanos;
-                }
-
-                @Override
-                public int getEndTimeTicks() {
-                    return stopTick;
-                }
-
-                @Override
-                public String getProfilerResults() {
-                    return "";
-                }
-            };
-        }
-    }
-
     // Paper start - Add tick times API and /mspt command
     public static class TickTimes {
         private final long[] times;
@@ -3164,24 +_,15 @@
 
     // Paper start - API to check if the server is sleeping
     public boolean isTickPaused() {
-        return this.emptyTicks > 0 && this.emptyTicks >= this.pauseWhenEmptySeconds() * 20;
+        return false; // Folia - region threading
     }
 
     public void addPluginAllowingSleep(final String pluginName, final boolean value) {
-        if (!value) {
-            this.pluginsBlockingSleep.add(pluginName);
-        } else {
-            this.pluginsBlockingSleep.remove(pluginName);
-        }
+        // Folia - region threading
     }
 
     private void removeDisabledPluginsBlockingSleep() {
-        if (this.pluginsBlockingSleep.isEmpty()) {
-            return;
-        }
-        this.pluginsBlockingSleep.removeIf(plugin -> (
-            !io.papermc.paper.plugin.manager.PaperPluginManagerImpl.getInstance().isPluginEnabled(plugin)
-        ));
+        // Folia - region threading
     }
     // Paper end - API to check if the server is sleeping
 }
