--- a/src/main/java/io/canvasmc/canvas/spark/FoliaPlatformInfo.java
+++ b/src/main/java/io/canvasmc/canvas/spark/FoliaPlatformInfo.java
@@ -14,8 +_,7 @@
     @Contract(pure = true)
     @Override
     public String getName() {
-        // TODO - should we change this to "Canvas"? We have an entirely custom implementation, may be worth it
-        return "Folia";
+        return "SourbyCraft";
     }
 
     @Override
@@ -24,7 +_,7 @@
 
     @Override
     public String getVersion() {
-        return this.server.getVersion();
+        return "Build44 (MC:" + this.server.getMinecraftVersion() + ")";
     }
 
     @Override
