From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sat, 6 Jun 2026 17:44:33 +0700
Subject: [PATCH] SourbyCraft: perf-engine P2 disable saving snowballs


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index d183a75c87626c680b96016c535f31acdfbc710e..0d80d5a9313c349d9dd83748cdbf74c953db98b8 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -67,4 +67,14 @@ public class Snowball extends ThrowableItemProjectile {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
+
+    // SourbyCraft start - perf-engine P2 lag-machine: skip NBT save for snowballs
+    @Override
+    public boolean shouldBeSaved() {
+        if (dev.iyanz.sourbycraft.perf.knob.Knobs.LAG_MACHINE_DISABLE_SAVING_SNOWBALLS.get()) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // SourbyCraft end - perf-engine P2 lag-machine
 }
