From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 20:30:37 +0200
Subject: [PATCH] Add CloudPlane configuration


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 2596601f9c5a3e726dd5c5777ad0c470f48eae07..8dc9cf41daa38f8647fce9ba2d70a445c08945b8 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -276,6 +276,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
         gg.pufferfish.pufferfish.PufferfishConfig.load(); // Pufferfish
         gg.pufferfish.pufferfish.PufferfishCommand.init(); // Pufferfish
+        dev.booky.cloudplane.CloudPlaneConfig.init((java.io.File) this.options.valueOf("cloudplane-settings")); // CloudPlane - config
 
         // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 703576ce396cb2ab8d36a58032ea1b07e2a7c922..dae5852d29b6f3b2515178cf259acd31c8e5a58a 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -818,6 +818,8 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
     // Paper end - optimise random ticking
 
+    public final dev.booky.cloudplane.CloudPlaneWorldConfig cloudplaneConfig; // CloudPlane - config
+
     protected Level(
         WritableLevelData levelData,
         ResourceKey<Level> dimension,
@@ -845,6 +847,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         // Paper end - getblock optimisations - cache world height/sections
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) levelData).getLevelName()); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
+        this.cloudplaneConfig = new dev.booky.cloudplane.CloudPlaneWorldConfig((((net.minecraft.world.level.storage.PrimaryLevelData) levelData).getLevelName()), environment); // CloudPlane - config
         this.generator = generator;
         this.world = new CraftWorld((ServerLevel) this, generator, biomeProvider, environment);
 
