From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: iyanz <boooky10@gmail.com>
Date: Tue, 24 Aug 2021 20:43:57 +0200
Subject: [PATCH] Add more detailed brand info

Sends version info to player in brand packet

diff --git a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index a3645322be006d5501fa7fb8944509a9c6b58e32..be6ece9040de523d94b94ff82dd342b9f5a2902c 100644
--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -101,7 +101,11 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
 
     public void startConfiguration() {
         new io.papermc.paper.event.connection.configuration.PlayerConnectionInitialConfigureEvent(this.paperConnection).callEvent(); // Paper
-        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
+        // SourbyCraft start - detailed brand info
+        String serverBrand = !dev.iyanz.sourbycraft.SourbyCraftConfig.detailedBrand ? this.server.getServerModName() :
+                this.server.getServerModName() + ' ' + io.papermc.paper.ServerBuildInfo.buildInfo().asString(io.papermc.paper.ServerBuildInfo.StringRepresentation.VERSION_SIMPLE);
+        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(serverBrand)));
+        // SourbyCraft end - detailed brand info
         ServerLinks serverLinks = this.server.serverLinks();
         if (!serverLinks.isEmpty()) {
             // Paper start
