From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sun, 14 Jun 2026 18:43:51 +0700
Subject: [PATCH] Fix redstone ore lit state desync on interact cancel


diff --git a/net/minecraft/world/level/block/RedStoneOreBlock.java b/net/minecraft/world/level/block/RedStoneOreBlock.java
index 60cf974d7b8b1ede092e2aa062083664388e07cc..5cb9208f7a1b56fede4dae431be6cee14b4cc360 100644
--- a/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -49,6 +49,8 @@ public class RedStoneOreBlock extends Block {
                 org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent(player, org.bukkit.event.block.Action.PHYSICAL, pos, null, null, null);
                 if (!event.isCancelled()) {
                     interact(level.getBlockState(pos), level, pos, entity); // add entity
+                } else if (entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) { // SourbyCraft - fix lit state desync on cancel
+                    serverPlayer.connection.send(new net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket(level, pos)); // SourbyCraft - fix lit state desync on cancel
                 }
             } else {
                 org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos));
