From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sun, 14 Jun 2026 18:44:49 +0700
Subject: [PATCH] Prevent healing amplifier from overflowing


diff --git a/net/minecraft/world/effect/HealOrHarmMobEffect.java b/net/minecraft/world/effect/HealOrHarmMobEffect.java
index e5a3f9f5bc760ed4ce0e6e721ffed25f6cdae9fe..93156aa44d86e0313d939c015805705f78d3e933 100644
--- a/net/minecraft/world/effect/HealOrHarmMobEffect.java
+++ b/net/minecraft/world/effect/HealOrHarmMobEffect.java
@@ -18,7 +18,7 @@ class HealOrHarmMobEffect extends InstantenousMobEffect {
         if (this.isHarm == mob.isInvertedHealAndHarm()) {
             mob.heal(Math.max(4 << amplification, 0), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
-            mob.hurtServer(level, mob.damageSources().magic(), 6 << amplification);
+            mob.hurtServer(level, mob.damageSources().magic(), Math.max(6 << amplification, 0)); // SourbyCraft - prevent negative healing/harming
         }
 
         return true;
@@ -36,13 +36,13 @@ class HealOrHarmMobEffect extends InstantenousMobEffect {
         if (!new io.papermc.paper.event.entity.EntityEffectTickEvent(mob.getBukkitLivingEntity(), org.bukkit.craftbukkit.potion.CraftPotionEffectType.minecraftToBukkit(this), amplification).callEvent()) { return; } // Paper - Add EntityEffectTickEvent
         if (this.isHarm == mob.isInvertedHealAndHarm()) {
             int amount = (int)(scale * (4 << amplification) + 0.5);
-            mob.heal(amount, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
+            mob.heal(Math.max(amount, 0), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit // SourbyCraft - prevent negative healing/harming
         } else {
             int amount = (int)(scale * (6 << amplification) + 0.5);
             if (source == null) {
-                mob.hurtServer(serverLevel, mob.damageSources().magic(), amount);
+                mob.hurtServer(serverLevel, mob.damageSources().magic(), Math.max(amount, 0)); // SourbyCraft - prevent negative healing/harming
             } else {
-                mob.hurtServer(serverLevel, mob.damageSources().indirectMagic(source, owner), amount);
+                mob.hurtServer(serverLevel, mob.damageSources().indirectMagic(source, owner), Math.max(amount, 0)); // SourbyCraft - prevent negative healing/harming
             }
         }
     }
