From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 5 Jun 2022 17:07:27 +0200
Subject: [PATCH] Log exceptions caused by packet sending


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index f6dec363af8834340356b3be936fc35862933b4e..2a6c58fb012acc3cf3f553ff10d8f646cca97de6 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -204,7 +204,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                     if (player != null) player.quitReason = org.bukkit.event.player.PlayerQuitEvent.QuitReason.ERRONEOUS_STATE; // Paper - Add API for quit reason
                     if (flag) {
-                        LOGGER.debug("Failed to sent packet", exception);
+                        LOGGER.error("Failed to sent packet", exception); // CloudPlane - always log errors
                         boolean doesDisconnectExist = this.packetListener.protocol() != ConnectionProtocol.STATUS && this.packetListener.protocol() != ConnectionProtocol.HANDSHAKING; // Paper
                         if (this.getSending() == PacketFlow.CLIENTBOUND && doesDisconnectExist) { // Paper
                             Packet<?> packet = (Packet<?>)(this.sendLoginDisconnect
@@ -217,7 +217,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                         this.setReadOnly();
                     } else {
-                        LOGGER.debug("Double fault", exception);
+                        LOGGER.error("Double fault", exception); // CloudPlane - always log errors
                         this.disconnect(disconnectionDetails);
                     }
                 }
