From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sun, 14 Jun 2026 18:56:59 +0700
Subject: [PATCH] Fire PlayerPickupArrowEvent for creative players


diff --git a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
index d25d525bad6b8096d80da19f3f969c1be514b115..932a8efbbaff157c7f2bd954577772683439f91a 100644
--- a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
@@ -697,7 +697,7 @@ public abstract class AbstractArrow extends Projectile {
         if (!this.level().isClientSide() && (this.isInGround() || this.isNoPhysics()) && this.shakeTime <= 0) {
             // CraftBukkit start
             ItemStack pickupItem = this.getPickupItem();
-            if (this.pickup == Pickup.ALLOWED && !pickupItem.isEmpty() && player.getInventory().canHold(pickupItem) > 0) {
+            if (!pickupItem.isEmpty() && ((this.pickup == Pickup.ALLOWED && player.getInventory().canHold(pickupItem) > 0) || (this.pickup == AbstractArrow.Pickup.CREATIVE_ONLY && player.getAbilities().instabuild))) { // SourbyCraft - fire arrow pickup for creative players
                 net.minecraft.world.entity.item.ItemEntity item = new net.minecraft.world.entity.item.ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), pickupItem);
                 org.bukkit.event.player.PlayerPickupArrowEvent event = new org.bukkit.event.player.PlayerPickupArrowEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.Item) item.getBukkitEntity(), (org.bukkit.entity.AbstractArrow) this.getBukkitEntity());
                 // event.setCancelled(!player.canPickUpLoot()); TODO
