From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sun, 14 Jun 2026 18:56:59 +0700
Subject: [PATCH] Log exceptions caused by packet sending


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 01e5ff70c70dd7fab089d8965bd1c237fa239f45..c4c4d2eeae473c35abbefd857971cb97a0440712 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -189,7 +189,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                     if (player != null) player.quitReason = org.bukkit.event.player.PlayerQuitEvent.QuitReason.ERRONEOUS_STATE; // Paper - Add API for quit reason
                     if (isFirstFault) {
-                        LOGGER.debug("Failed to sent packet", cause);
+                        LOGGER.error("Failed to sent packet", cause); // SourbyCraft - always log errors
                         boolean doesDisconnectExist = this.packetListener.protocol() != ConnectionProtocol.STATUS && this.packetListener.protocol() != ConnectionProtocol.HANDSHAKING; // Paper
                         if (this.getSending() == PacketFlow.CLIENTBOUND && doesDisconnectExist) { // Paper
                             Packet<?> packet = this.sendLoginDisconnect
@@ -202,7 +202,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                         this.setReadOnly();
                     } else {
-                        LOGGER.debug("Double fault", cause);
+                        LOGGER.error("Double fault", cause); // SourbyCraft - always log errors
                         this.disconnect(details);
                     }
                 }
