From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: iyanz <boooky10@gmail.com>
Date: Mon, 27 May 2024 03:13:37 +0200
Subject: [PATCH] Add option to translate custom item names and lore


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index fb2cd5fe354ad8f024e3f0d7a1f28eda4b2331f6..7dd40c809843bf34df6f4fb8b1bfdd06825a2856 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -196,7 +196,13 @@ public final class ItemStack implements DataComponentHolder {
                 } else {
                     Holder<Item> holder = Item.STREAM_CODEC.decode(buffer);
                     DataComponentPatch dataComponentPatch = codec.decode(buffer);
-                    return new ItemStack(holder, varInt, dataComponentPatch);
+                    // SourbyCraft start - item localization
+                    ItemStack stack = new ItemStack(holder, varInt, dataComponentPatch);
+                    if (dev.iyanz.sourbycraft.SourbyCraftConfig.localizeItems) {
+                        dev.iyanz.sourbycraft.ItemUtil.unpackPatchSaves(stack);
+                    }
+                    return stack;
+                    // SourbyCraft end - item localization
                 }
             }
 
@@ -207,15 +213,23 @@ public final class ItemStack implements DataComponentHolder {
                 } else {
                     buffer.writeVarInt(io.papermc.paper.util.sanitizer.ItemComponentSanitizer.sanitizeCount(io.papermc.paper.util.sanitizer.ItemObfuscationSession.currentSession(), value, value.getCount())); // Paper - potentially sanitize count
                     Item.STREAM_CODEC.encode(buffer, value.getItemHolder());
+                    // SourbyCraft start - item localization
+                    if (dev.iyanz.sourbycraft.SourbyCraftConfig.localizeItems && buffer.adventure$locale != null) {
+                        try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(value))) { // pass the itemstack as context to the obfuscation session
+                            codec.encode(buffer, dev.iyanz.sourbycraft.ItemUtil.packPatchSaves(value));
+                        }
+                    } else {
+                        // SourbyCraft end - item localization
                     // Paper start - adventure; conditionally render translatable components
                     boolean prev = net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
-                    try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(value))) { // pass the itemstack as context to the obfuscation session
+                    try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(value))) { // pass the itemstack as context to the obfuscation session // SourbyCraft - item localization; diff on change
                         net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(true);
-                    codec.encode(buffer, value.components.asPatch());
+                    codec.encode(buffer, value.components.asPatch()); // SourbyCraft - item localization; diff on change
                     } finally {
                         net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prev);
                     }
                     // Paper end - adventure; conditionally render translatable components
+                    } // SourbyCraft - item localization
                 }
             }
         };
diff --git a/net/minecraft/world/item/component/ItemLore.java b/net/minecraft/world/item/component/ItemLore.java
index a279e04868ed6df1aeda07e5fef8460458ec24af..bf954b0372737cf51ed72220ad1e727a08e04fe1 100644
--- a/net/minecraft/world/item/component/ItemLore.java
+++ b/net/minecraft/world/item/component/ItemLore.java
@@ -42,6 +42,36 @@ public record ItemLore(List<Component> lines, List<Component> styledLines) imple
                 }
                 return realLines;
             })
+            // SourbyCraft start - item localization; line splitting support
+            .<List<Component>>apply(codec -> new StreamCodec<>() {
+                @Override
+                public List<Component> decode(RegistryFriendlyByteBuf buf) {
+                    return codec.decode(buf);
+                }
+
+                @Override
+                public void encode(RegistryFriendlyByteBuf buf, List<Component> value) {
+                    if (dev.iyanz.sourbycraft.SourbyCraftConfig.localizeItems && buf.adventure$locale != null) {
+                        // pre-render translations here to work around line splitting not working anymore
+                        boolean prevNoRender = ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
+                        try {
+                            ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(true);
+                            List<Component> lines = Lists.newArrayListWithCapacity(value.size());
+                            for (Component line : value) {
+                                net.kyori.adventure.text.Component advLine = io.papermc.paper.adventure.PaperAdventure.asAdventure(line);
+                                net.kyori.adventure.text.Component i18nLine = net.kyori.adventure.translation.GlobalTranslator.render(advLine, buf.adventure$locale);
+                                lines.add(io.papermc.paper.adventure.PaperAdventure.asVanilla(i18nLine));
+                            }
+                            codec.encode(buf, lines);
+                        } finally {
+                            ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prevNoRender);
+                        }
+                    } else {
+                        codec.encode(buf, value);
+                    }
+                }
+            })
+            // SourbyCraft end - item localization; line splitting support
             .map(ItemLore::new, ItemLore::lines);
     // SourbyCraft end - split lore lines
 
diff --git a/net/minecraft/world/item/trading/ItemCost.java b/net/minecraft/world/item/trading/ItemCost.java
index 02cc7a23a0fb6a9e4ed88578606c40d5d2301ae1..fe41f833d72cdfd9de7329c72a7497b22d9b5db2 100644
--- a/net/minecraft/world/item/trading/ItemCost.java
+++ b/net/minecraft/world/item/trading/ItemCost.java
@@ -32,7 +32,7 @@ public record ItemCost(Holder<Item> item, int count, DataComponentExactPredicate
         DataComponentExactPredicate.STREAM_CODEC,
         ItemCost::components,
         ItemCost::new
-    );
+    ).map(dev.iyanz.sourbycraft.ItemUtil::unpackPatchSaves, dev.iyanz.sourbycraft.ItemUtil::packPatchSaves); // SourbyCraft - item translations; fix merchant offers de-sync
     public static final StreamCodec<RegistryFriendlyByteBuf, Optional<ItemCost>> OPTIONAL_STREAM_CODEC = STREAM_CODEC.apply(ByteBufCodecs::optional);
 
     public ItemCost(ItemLike item) {
