From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sun, 14 Jun 2026 18:56:59 +0700
Subject: [PATCH] Log exceptions caused by packet sending


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 0f06d509125e72ec11d263d0d5c2179d65f6d8dd..397b00a3f5306c785f344622154151cb74644971 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -188,7 +188,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                     if (player != null) player.quitReason = org.bukkit.event.player.PlayerQuitEvent.QuitReason.ERRONEOUS_STATE; // Paper - Add API for quit reason
                     if (isFirstFault) {
-                        LOGGER.debug("Failed to sent packet", cause);
+                        LOGGER.error("Failed to sent packet", cause); // SourbyCraft - always log errors
                         boolean doesDisconnectExist = this.packetListener.protocol() != ConnectionProtocol.STATUS && this.packetListener.protocol() != ConnectionProtocol.HANDSHAKING; // Paper
                         if (this.getSending() == PacketFlow.CLIENTBOUND && doesDisconnectExist) { // Paper
                             Packet<?> packet = this.sendLoginDisconnect
@@ -201,7 +201,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                         this.setReadOnly();
                     } else {
-                        LOGGER.debug("Double fault", cause);
+                        LOGGER.error("Double fault", cause); // SourbyCraft - always log errors
                         this.disconnect(details);
                     }
                 }
