From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: iyanz <boooky10@gmail.com>
Date: Sat, 31 Aug 2024 18:57:57 +0200
Subject: [PATCH] Fix elastic leash behaviour on unleash cancel


diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index 54c7af09f5f64021895074ea5a1578513aa67bf2..8122e21cb32e5ba5b2e14dfc6a6282206cd6804b 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -170,8 +170,7 @@ public interface Leashable {
             if (leashHolder != null && leashHolder.level() == entity.level()) {
                 double d = entity.leashDistanceTo(leashHolder);
                 entity.whenLeashedTo(leashHolder);
-                if (d > entity.leashSnapDistanceOrConfig()) { // Paper - Configurable max leash distance
-                    entity.leashTooFarBehaviour();
+                if (d > entity.leashSnapDistanceOrConfig() && entity.leashTooFarBehaviour0()) { // Paper - Configurable max leash distance // SourbyCraft - fix elastic leash behavior on event cancel
                 } else if (d > entity.leashElasticDistance() - leashHolder.getBbWidth() - entity.getBbWidth()
                     && entity.checkElasticInteractions(leashHolder, leashData)) {
                     entity.onElasticLeashPull();
@@ -216,13 +215,21 @@ public interface Leashable {
         entity.notifyLeashHolder(this);
     }
 
+    // SourbyCraft start - fix elastic leash behaviour on unleash cancel
+    @io.papermc.paper.annotation.DoNotUse
+    @Deprecated
     default void leashTooFarBehaviour() {
+        this.leashTooFarBehaviour0();
+    }
+
+    default boolean leashTooFarBehaviour0() {
+        // SourbyCraft end - fix elastic leash behaviour on unleash cancel
         // CraftBukkit start
         boolean dropLeash = true; // Paper
         if (this instanceof Entity entity) {
             // Paper start - Expand EntityUnleashEvent
             final org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(entity.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE, true);
-            if (!event.callEvent()) return;
+            if (!event.callEvent()) return false; // SourbyCraft - fix elastic leash behaviour on unleash cancel
 
             Entity leashHolder = this.getLeashHolder();
             Level level = leashHolder.level();
@@ -236,6 +243,7 @@ public interface Leashable {
             this.removeLeash();
         }
         // Paper end - Expand EntityUnleashEvent
+        return true; // SourbyCraft - fix elastic leash behaviour on unleash cancel
     }
 
     default void closeRangeLeashBehaviour(Entity entity) {
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 60e3c2d6d0e5abbc48c88e416f3adeae0e2f9806..6f7fcc90d9204a7738240c36a13e27efbe05fcc4 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1488,10 +1488,14 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         }
     }
 
+    // SourbyCraft start - fix elastic leash behaviour on unleash cancel
     @Override
-    public void leashTooFarBehaviour() {
-        Leashable.super.leashTooFarBehaviour();
+    public boolean leashTooFarBehaviour0() {
+        boolean ret = Leashable.super.leashTooFarBehaviour0();
+        if (ret)
         this.goalSelector.disableControlFlag(Goal.Flag.MOVE);
+        return ret;
+        // SourbyCraft end - fix elastic leash behaviour on unleash cancel
     }
 
     @Override
