From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 20 Sep 2024 00:56:09 +0200
Subject: [PATCH] Work around unnecessary chunk pos conversion on entity
 tracker visibility check


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 7129c5e1920008ac54f3a8ac83f5589396f4e4e9..a692729183194a36c9a042ccf625e8591ca331f1 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -273,18 +273,26 @@ public final class RegionizedPlayerChunkLoader {
         return sendViewDistance;
     }
 
+    // CloudPlane start - work around unnecessary chunk pos conversion
+    @Deprecated
+    @io.papermc.paper.annotation.DoNotUse
     public boolean isChunkSent(final ServerPlayer player, final int chunkX, final int chunkZ, final boolean borderOnly) {
         return borderOnly ? this.isChunkSentBorderOnly(player, chunkX, chunkZ) : this.isChunkSent(player, chunkX, chunkZ);
     }
 
     public boolean isChunkSent(final ServerPlayer player, final int chunkX, final int chunkZ) {
+        return this.isChunkSent(player, CoordinateUtils.getChunkKey(chunkX, chunkZ));
+    }
+
+    public boolean isChunkSent(final ServerPlayer player, final long chunkKey) {
         final PlayerChunkLoaderData loader = ((ChunkSystemServerPlayer)player).moonrise$getChunkLoader();
         if (loader == null) {
             return false;
         }
 
-        return loader.sentChunks.contains(CoordinateUtils.getChunkKey(chunkX, chunkZ));
+        return loader.sentChunks.contains(chunkKey);
     }
+    // CloudPlane end - work around unnecessary chunk pos conversion
 
     public boolean isChunkSentBorderOnly(final ServerPlayer player, final int chunkX, final int chunkZ) {
         final PlayerChunkLoaderData loader = ((ChunkSystemServerPlayer)player).moonrise$getChunkLoader();
diff --git a/net/minecraft/server/level/ChunkHolder.java b/net/minecraft/server/level/ChunkHolder.java
index 22123d8b0fd8741c6df4efaec56f3c694e0529a3..66ccb5eb9784b7502abfba4fdbbeeb859f1e52cd 100644
--- a/net/minecraft/server/level/ChunkHolder.java
+++ b/net/minecraft/server/level/ChunkHolder.java
@@ -93,7 +93,7 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
         final ServerPlayer[] raw = this.playersSentChunkTo.getRawDataUnchecked();
         for (int i = 0, len = this.playersSentChunkTo.size(); i < len; ++i) {
             final ServerPlayer player = raw[i];
-            if (onlyOnWatchDistanceEdge && !((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.getChunkMap().level).moonrise$getPlayerChunkLoader().isChunkSent(player, this.pos.x, this.pos.z, onlyOnWatchDistanceEdge)) {
+            if (onlyOnWatchDistanceEdge && !((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.getChunkMap().level).moonrise$getPlayerChunkLoader().isChunkSentBorderOnly(player, this.pos.x, this.pos.z)) { // CloudPlane - work around unnecessary chunk pos conversion
                 continue;
             }
             ret.add(player);
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index ab8893163ac85971a470e4a4d6fa76bd174e2bae..fe8d8d0bba4bd66a3cc203f1d1cad559cdd701d5 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -305,7 +305,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
     }
 
     private boolean isChunkOnTrackedBorder(ServerPlayer player, int x, int z) {
-        return ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getPlayerChunkLoader().isChunkSent(player, x, z, true); // Paper - rewrite chunk system
+        return ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getPlayerChunkLoader().isChunkSentBorderOnly(player, x, z); // Paper - rewrite chunk system // CloudPlane - work around unnecessary chunk pos conversion
     }
 
     protected ThreadedLevelLightEngine getLightEngine() {
