From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sun, 14 Jun 2026 23:49:59 +0700
Subject: [PATCH] SourbyCraft US import: sounds disablePiglinAngerSound gate


diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index 77638e1b9b2ab1165d11c1ac8ea09255b04544f6..6759a2714981d8c04e71a7d36c303a305b5eab4c 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -428,7 +428,10 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
 
     @Override
     public @Nullable SoundEvent getAmbientSound() {
-        return this.level().isClientSide() ? null : PiglinAi.getSoundForCurrentActivity(this).orElse(null);
+        if (this.level().isClientSide()) return null; // SourbyCraft - US import
+        SoundEvent s = PiglinAi.getSoundForCurrentActivity(this).orElse(null); // SourbyCraft - US import
+        if (s == SoundEvents.PIGLIN_ANGRY && dev.iyanz.sourbycraft.SourbyCraftConfig.ymlBool("sounds.disablePiglinAngerSound", false)) return null; // SourbyCraft - US import
+        return s; // SourbyCraft - US import
     }
 
     @Override
