From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sun, 14 Jun 2026 23:49:59 +0700
Subject: [PATCH] SourbyCraft US import: sounds disablePiglinAngerSound gate


diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index b0e8b2db95aa5a0f53ddf8c7d082963dc693f879..88d69697c0c3ce4ed555069fc256f7f44df5c5ee 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -429,7 +429,10 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
 
     @Override
     public @Nullable SoundEvent getAmbientSound() {
-        return this.level().isClientSide() ? null : PiglinAi.getSoundForCurrentActivity(this).orElse(null);
+        if (this.level().isClientSide()) return null; // SourbyCraft - US import
+        SoundEvent s = PiglinAi.getSoundForCurrentActivity(this).orElse(null); // SourbyCraft - US import
+        if (s == SoundEvents.PIGLIN_ANGRY && dev.iyanz.sourbycraft.SourbyCraftConfig.ymlBool("sounds.disablePiglinAngerSound", false)) return null; // SourbyCraft - US import
+        return s; // SourbyCraft - US import
     }
 
     @Override
