From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: iyanz <boooky10@gmail.com>
Date: Sun, 5 Jan 2025 21:28:12 +0100
Subject: [PATCH] Convert chunk heightmaps to array


diff --git a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index 19547e22620f16d01f05bbd13a4dcb9c0df636df..af940c5862aee17b058f19699d7064be7a534709 100644
--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -28,8 +28,17 @@ public class ClientboundLevelChunkPacketData {
     private static final StreamCodec<ByteBuf, Map<Heightmap.Types, long[]>> HEIGHTMAPS_STREAM_CODEC = ByteBufCodecs.map(
         i -> new EnumMap<>(Heightmap.Types.class), Heightmap.Types.STREAM_CODEC, ByteBufCodecs.LONG_ARRAY
     );
+    // SourbyCraft start - convert heightmaps to array
+    public static final Heightmap.Types[] SENDABLE_HEIGHTMAP_TYPES = java.util.Arrays.stream(Heightmap.Types.values())
+            .filter(Heightmap.Types::sendToClient).toArray(Heightmap.Types[]::new);
+    public static final int[] SENDABLE_HEIGHTMAP_TYPE_IDS = java.util.Arrays.stream(Heightmap.Types.values())
+            .filter(Heightmap.Types::sendToClient).mapToInt(Enum::ordinal).toArray();
+    public static final int  SENDABLE_HEIGHTMAP_TYPES_LEN = SENDABLE_HEIGHTMAP_TYPES.length;
+    public static final int HEIGHTMAP_TYPES_LEN = Heightmap.Types.values().length;
     private static final int TWO_MEGABYTES = 2097152;
-    private final Map<Heightmap.Types, long[]> heightmaps;
+    private final long[][] heightmaps;
+    private final int heightmapsCount;
+    // SourbyCraft end - convert heightmaps to array
     private final byte[] buffer;
     private final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData;
     // Paper start - Handle oversized block entities in chunks
@@ -48,10 +57,19 @@ public class ClientboundLevelChunkPacketData {
     }
     public ClientboundLevelChunkPacketData(LevelChunk levelChunk, io.papermc.paper.antixray.ChunkPacketInfo<net.minecraft.world.level.block.state.BlockState> chunkPacketInfo) {
         // Paper end - Anti-Xray - Add chunk packet info
-        this.heightmaps = levelChunk.getHeightmaps()
-            .stream()
-            .filter(entry1 -> entry1.getKey().sendToClient())
-            .collect(Collectors.toMap(Entry::getKey, entry1 -> (long[])entry1.getValue().getRawData().clone()));
+        // SourbyCraft start - convert heightmaps to array
+        this.heightmaps = new long[HEIGHTMAP_TYPES_LEN][];
+        int heightmapsCount = 0;
+        for (int i = 0; i < SENDABLE_HEIGHTMAP_TYPES_LEN; ++i) {
+            int typeId = SENDABLE_HEIGHTMAP_TYPE_IDS[i];
+            Heightmap heightmap = levelChunk.heightmaps[typeId];
+            if (heightmap != null) {
+                this.heightmaps[typeId] = heightmap.getRawData();
+                heightmapsCount++;
+            }
+        }
+        this.heightmapsCount = heightmapsCount;
+        // SourbyCraft end - convert heightmaps to array
         this.buffer = new byte[calculateChunkSize(levelChunk)];
         // Paper start - Anti-Xray - Add chunk packet info
         if (chunkPacketInfo != null) {
@@ -76,7 +94,13 @@ public class ClientboundLevelChunkPacketData {
     }
 
     public ClientboundLevelChunkPacketData(RegistryFriendlyByteBuf buffer, int x, int z) {
-        this.heightmaps = HEIGHTMAPS_STREAM_CODEC.decode(buffer);
+        // SourbyCraft start - convert heightmaps to array
+        this.heightmaps = new long[HEIGHTMAP_TYPES_LEN][];
+        this.heightmapsCount = ByteBufCodecs.readCount(buffer, HEIGHTMAP_TYPES_LEN);
+        for (int i = 0; i < this.heightmapsCount; i++) {
+            this.heightmaps[net.minecraft.network.VarInt.read(buffer)] = buffer.readLongArray();
+        }
+        // SourbyCraft end - convert heightmaps to array
         int varInt = buffer.readVarInt();
         if (varInt > 2097152) { // Paper - diff on change - if this changes, update PacketEncoder
             throw new RuntimeException("Chunk Packet trying to allocate too much memory on read.");
@@ -88,7 +112,16 @@ public class ClientboundLevelChunkPacketData {
     }
 
     public void write(RegistryFriendlyByteBuf buffer) {
-        HEIGHTMAPS_STREAM_CODEC.encode(buffer, this.heightmaps);
+        // SourbyCraft start - convert heightmaps to array
+        ByteBufCodecs.writeCount(buffer, this.heightmapsCount, HEIGHTMAP_TYPES_LEN);
+        for (int i = 0; i < this.heightmaps.length; i++) {
+            long[] data = this.heightmaps[i];
+            if (data != null) {
+                net.minecraft.network.VarInt.write(buffer, i);
+                buffer.writeLongArray(data);
+            }
+        }
+        // SourbyCraft end - convert heightmaps to array
         buffer.writeVarInt(this.buffer.length);
         buffer.writeBytes(this.buffer);
         ClientboundLevelChunkPacketData.BlockEntityInfo.LIST_STREAM_CODEC.encode(buffer, this.blockEntitiesData);
@@ -150,7 +183,7 @@ public class ClientboundLevelChunkPacketData {
     }
 
     public Map<Heightmap.Types, long[]> getHeightmaps() {
-        return this.heightmaps;
+        throw new UnsupportedOperationException(); // SourbyCraft - convert heightmaps to array
     }
 
     static class BlockEntityInfo {
diff --git a/net/minecraft/world/level/chunk/ChunkAccess.java b/net/minecraft/world/level/chunk/ChunkAccess.java
index 54ddcf92e72b9cbd0eb442e5d0faa83c0cf3b33b..484319bbf9bb3eb9c4bdfb69364ce93cc9cfcc42 100644
--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -71,7 +71,7 @@ public abstract class ChunkAccess implements BiomeManager.NoiseBiomeSource, Ligh
     protected @Nullable NoiseChunk noiseChunk;
     protected final UpgradeData upgradeData;
     protected @Nullable BlendingData blendingData;
-    public final Map<Heightmap.Types, Heightmap> heightmaps = Maps.newEnumMap(Heightmap.Types.class);
+    public final @Nullable Heightmap[] heightmaps = new Heightmap[Heightmap.Types.values().length]; // SourbyCraft - convert heightmaps to array
     // Paper - rewrite chunk system
     private final Map<Structure, StructureStart> structureStarts = Maps.newHashMap();
     private final Map<Structure, LongSet> structuresRefences = Maps.newHashMap();
@@ -231,31 +231,42 @@ public abstract class ChunkAccess implements BiomeManager.NoiseBiomeSource, Ligh
         return this.getSections()[index];
     }
 
+    // SourbyCraft start - convert heightmaps to array
+    @io.papermc.paper.annotation.DoNotUse
+    @Deprecated
     public Collection<Entry<Heightmap.Types, Heightmap>> getHeightmaps() {
-        return Collections.unmodifiableSet(this.heightmaps.entrySet());
+        throw new UnsupportedOperationException();
     }
+    // SourbyCraft end - convert heightmaps to array
 
     public void setHeightmap(Heightmap.Types type, long[] data) {
         this.getOrCreateHeightmapUnprimed(type).setRawData(this, type, data);
     }
 
     public Heightmap getOrCreateHeightmapUnprimed(Heightmap.Types type) {
-        return this.heightmaps.computeIfAbsent(type, absentType -> new Heightmap(this, absentType));
+        // SourbyCraft start - convert heightmaps to array
+        Heightmap heightmap = this.heightmaps[type.ordinal()];
+        if (heightmap == null) {
+            heightmap = new Heightmap(this, type);
+            this.heightmaps[type.ordinal()] = heightmap;
+        }
+        return heightmap;
+        // SourbyCraft end - convert heightmaps to array
     }
 
     public boolean hasPrimedHeightmap(Heightmap.Types type) {
-        return this.heightmaps.get(type) != null;
+        return this.heightmaps[type.ordinal()] != null; // SourbyCraft - convert heightmaps to array
     }
 
     public int getHeight(Heightmap.Types type, int x, int z) {
-        Heightmap heightmap = this.heightmaps.get(type);
+        Heightmap heightmap = this.heightmaps[type.ordinal()]; // SourbyCraft - convert heightmaps to array
         if (heightmap == null) {
             if (SharedConstants.IS_RUNNING_IN_IDE && this instanceof LevelChunk) {
                 LOGGER.error("Unprimed heightmap: {} {} {}", type, x, z);
             }
 
             Heightmap.primeHeightmaps(this, EnumSet.of(type));
-            heightmap = this.heightmaps.get(type);
+            heightmap = this.heightmaps[type.ordinal()]; // SourbyCraft - convert heightmaps to array
         }
 
         return heightmap.getFirstAvailable(x & 15, z & 15) - 1;
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 79b8d7e5de161944b5b2308c0a25cd9c17f26ee6..7efc5498fca6e1aa86342fb4f6ef9247c001426b 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -168,11 +168,11 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         this.level = (ServerLevel) level; // CraftBukkit - type
         this.gameEventListenerRegistrySections = new Int2ObjectOpenHashMap<>();
 
-        for (Heightmap.Types types : Heightmap.Types.values()) {
-            if (ChunkStatus.FULL.heightmapsAfter().contains(types)) {
-                this.heightmaps.put(types, new Heightmap(this, types));
-            }
+        // SourbyCraft start - convert heightmaps to array
+        for (Heightmap.Types types : ChunkStatus.FULL.heightmapsAfter()) {
+            this.heightmaps[types.ordinal()] = new Heightmap(this, types);
         }
+        // SourbyCraft end - convert heightmaps to array
 
         this.postLoad = postLoad;
         this.blockTicks = blockTicks;
@@ -218,11 +218,14 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         this.setAllStarts(chunk.getAllStarts());
         this.setAllReferences(chunk.getAllReferences());
 
-        for (Entry<Heightmap.Types, Heightmap> entry : chunk.getHeightmaps()) {
-            if (ChunkStatus.FULL.heightmapsAfter().contains(entry.getKey())) {
-                this.setHeightmap(entry.getKey(), entry.getValue().getRawData());
+        // SourbyCraft start - convert heightmaps to array
+        for (Heightmap.Types type : ChunkStatus.FULL.heightmapsAfter()) {
+            Heightmap heightmap = chunk.heightmaps[type.ordinal()];
+            if (heightmap != null) {
+                this.setHeightmap(type, heightmap.getRawData());
             }
         }
+        // SourbyCraft end - convert heightmaps to array
 
         // Paper - rewrite chunk system
         this.setLightCorrect(chunk.isLightCorrect());
@@ -394,10 +397,10 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                 return null;
             } else {
                 Block block = state.getBlock();
-                this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.OCEAN_FLOOR).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.WORLD_SURFACE).update(i, y, i2, state);
+                this.heightmaps[Heightmap.Types.MOTION_BLOCKING.ordinal()].update(i, y, i2, state); // SourbyCraft - convert heightmaps to array
+                this.heightmaps[Heightmap.Types.MOTION_BLOCKING_NO_LEAVES.ordinal()].update(i, y, i2, state); // SourbyCraft - convert heightmaps to array
+                this.heightmaps[Heightmap.Types.OCEAN_FLOOR.ordinal()].update(i, y, i2, state); // SourbyCraft - convert heightmaps to array
+                this.heightmaps[Heightmap.Types.WORLD_SURFACE.ordinal()].update(i, y, i2, state); // SourbyCraft - convert heightmaps to array
                 boolean hasOnlyAir1 = section.hasOnlyAir();
                 if (hasOnlyAir != hasOnlyAir1) {
                     this.level.getChunkSource().getLightEngine().updateSectionStatus(pos, hasOnlyAir1);
diff --git a/net/minecraft/world/level/chunk/ProtoChunk.java b/net/minecraft/world/level/chunk/ProtoChunk.java
index fd923d89d4456f0b13d97ef6fa90815a27e2e734..c37792ac5f2f9d15cf80076ddf72ed8ad1db7de7 100644
--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -163,7 +163,7 @@ public class ProtoChunk extends ChunkAccess {
                 EnumSet<Heightmap.Types> set1 = null;
 
                 for (Heightmap.Types types : set) {
-                    Heightmap heightmap = this.heightmaps.get(types);
+                    Heightmap heightmap = this.heightmaps[types.ordinal()]; // SourbyCraft - convert heightmaps to array
                     if (heightmap == null) {
                         if (set1 == null) {
                             set1 = EnumSet.noneOf(Heightmap.Types.class);
@@ -178,7 +178,7 @@ public class ProtoChunk extends ChunkAccess {
                 }
 
                 for (Heightmap.Types typesx : set) {
-                    this.heightmaps.get(typesx).update(relativeBlockPosX, y, relativeBlockPosZ, state);
+                    this.heightmaps[typesx.ordinal()].update(relativeBlockPosX, y, relativeBlockPosZ, state); // SourbyCraft - convert heightmaps to array
                 }
 
                 return blockState;
diff --git a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
index 95c8d65a4b6db8d0721460ae55cb8fb8143ea9a2..3043e080523ef07a83ce8db8f4ca472bac030ce6 100644
--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -518,12 +518,15 @@ public record SerializableChunkData(
 
             Map<Heightmap.Types, long[]> map = new EnumMap<>(Heightmap.Types.class);
 
-            for (Entry<Heightmap.Types, Heightmap> entry : chunk.getHeightmaps()) {
-                if (chunk.getPersistedStatus().heightmapsAfter().contains(entry.getKey())) {
-                    long[] rawData = entry.getValue().getRawData();
-                    map.put(entry.getKey(), (long[])rawData.clone());
+            // SourbyCraft start - convert heightmaps to array
+            for (Heightmap.Types type : chunk.getPersistedStatus().heightmapsAfter()) {
+                Heightmap heightmap = chunk.heightmaps[type.ordinal()];
+                if (heightmap != null) {
+                    long[] rawData = heightmap.getRawData();
+                    map.put(type, (long[])rawData.clone());
                 }
             }
+            // SourbyCraft end - convert heightmaps to array
 
             ChunkAccess.PackedTicks ticksForSerialization = chunk.getTicksForSerialization(level.getGameTime());
             ShortList[] lists = Arrays.stream(chunk.getPostProcessing())
