From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Mon, 15 Jun 2026 15:27:04 +0700
Subject: [PATCH] Fix elastic leash behaviour on unleash cancel


diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index 98034a4c96bf2972316078d3b3a49140921aab50..e91ce207924b9ddadb070447964c6171015a7cab 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -167,8 +167,7 @@ public interface Leashable {
             if (leashHolder != null && leashHolder.level() == entity.level()) {
                 double distanceTo = entity.leashDistanceTo(leashHolder);
                 entity.whenLeashedTo(leashHolder);
-                if (distanceTo > entity.leashSnapDistanceOrConfig()) { // Paper - Configurable max leash distance
-                    entity.leashTooFarBehaviour();
+                if (distanceTo > entity.leashSnapDistanceOrConfig() && entity.leashTooFarBehaviour0()) { // Paper - Configurable max leash distance // SourbyCraft - fix elastic leash behavior on event cancel
                 } else if (distanceTo > entity.leashElasticDistance() - leashHolder.getBbWidth() - entity.getBbWidth()
                     && entity.checkElasticInteractions(leashHolder, leashData)) {
                     entity.onElasticLeashPull();
@@ -213,7 +212,15 @@ public interface Leashable {
         leashHolder.notifyLeashHolder(this);
     }
 
+    // SourbyCraft start - fix elastic leash behaviour on unleash cancel
+    @io.papermc.paper.annotation.DoNotUse
+    @Deprecated
     default void leashTooFarBehaviour() {
+        this.leashTooFarBehaviour0();
+    }
+
+    default boolean leashTooFarBehaviour0() {
+        // SourbyCraft end - fix elastic leash behaviour on unleash cancel
         // CraftBukkit start
         boolean dropLeash = true; // Paper
         if (this instanceof Entity entity) {
@@ -222,7 +229,7 @@ public interface Leashable {
             Level level = leashHolder.level();
 
             final org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(entity.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE, true);
-            if (!event.callEvent()) return;
+            if (!event.callEvent()) return false; // SourbyCraft - fix elastic leash behaviour on unleash cancel
 
             dropLeash = event.isDropLeash();
             level.playSound(null, leashHolder.getX(), leashHolder.getY(), leashHolder.getZ(), SoundEvents.LEAD_BREAK, SoundSource.NEUTRAL, 1.0F, 1.0F); // Moved from Leashable#tickLeash
@@ -234,6 +241,7 @@ public interface Leashable {
             this.removeLeash();
         }
         // Paper end - Expand EntityUnleashEvent
+        return true; // SourbyCraft - fix elastic leash behaviour on unleash cancel
     }
 
     default void closeRangeLeashBehaviour(final Entity leashHolder) {
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index c2dd2da4bfb7d65ad2730ba1ce771e2b488beda4..79e6966be8394b96317e7f8e6031a12627038953 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1497,10 +1497,14 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         }
     }
 
+    // SourbyCraft start - fix elastic leash behaviour on unleash cancel
     @Override
-    public void leashTooFarBehaviour() {
-        Leashable.super.leashTooFarBehaviour();
+    public boolean leashTooFarBehaviour0() {
+        boolean ret = Leashable.super.leashTooFarBehaviour0();
+        if (ret)
         this.goalSelector.disableControlFlag(Goal.Flag.MOVE);
+        return ret;
+        // SourbyCraft end - fix elastic leash behaviour on unleash cancel
     }
 
     @Override
