From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Mon, 15 Jun 2026 15:27:04 +0700
Subject: [PATCH] Fix elastic leash behaviour on unleash cancel


diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index 132ce6cef9f9f485028246c41bee805d5d0fe1ec..29f8d9d403d6fbbac2bf649bf3a556773096393e 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -171,8 +171,7 @@ public interface Leashable {
             if (leashHolder != null && leashHolder.level() == entity.level()) {
                 double distanceTo = entity.leashDistanceTo(leashHolder);
                 entity.whenLeashedTo(leashHolder);
-                if (distanceTo > entity.leashSnapDistanceOrConfig()) { // Paper - Configurable max leash distance
-                    entity.leashTooFarBehaviour();
+                if (distanceTo > entity.leashSnapDistanceOrConfig() && entity.leashTooFarBehaviour0()) { // Paper - Configurable max leash distance // SourbyCraft - fix elastic leash behavior on event cancel
                 } else if (distanceTo > entity.leashElasticDistance() - leashHolder.getBbWidth() - entity.getBbWidth()
                     && entity.checkElasticInteractions(leashHolder, leashData)) {
                     entity.onElasticLeashPull();
@@ -217,13 +216,21 @@ public interface Leashable {
         leashHolder.notifyLeashHolder(this);
     }
 
+    // SourbyCraft start - fix elastic leash behaviour on unleash cancel
+    @io.papermc.paper.annotation.DoNotUse
+    @Deprecated
     default void leashTooFarBehaviour() {
+        this.leashTooFarBehaviour0();
+    }
+
+    default boolean leashTooFarBehaviour0() {
+        // SourbyCraft end - fix elastic leash behaviour on unleash cancel
         // CraftBukkit start
         boolean dropLeash = true; // Paper
         if (this instanceof Entity entity) {
             // Paper start - Expand EntityUnleashEvent
             final org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(entity.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE, true);
-            if (!event.callEvent()) return;
+            if (!event.callEvent()) return false; // SourbyCraft - fix elastic leash behaviour on unleash cancel
 
             Entity leashHolder = this.getLeashHolder();
             Level level = leashHolder.level();
@@ -237,6 +244,7 @@ public interface Leashable {
             this.removeLeash();
         }
         // Paper end - Expand EntityUnleashEvent
+        return true; // SourbyCraft - fix elastic leash behaviour on unleash cancel
     }
 
     default void closeRangeLeashBehaviour(final Entity leashHolder) {
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index ea144213f5ac2f813c9007113e4066a3295a7f76..cb4260791501c0c15c5a7b0509d6631fefb492b7 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1485,10 +1485,14 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         }
     }
 
+    // SourbyCraft start - fix elastic leash behaviour on unleash cancel
     @Override
-    public void leashTooFarBehaviour() {
-        Leashable.super.leashTooFarBehaviour();
+    public boolean leashTooFarBehaviour0() {
+        boolean ret = Leashable.super.leashTooFarBehaviour0();
+        if (ret)
         this.goalSelector.disableControlFlag(Goal.Flag.MOVE);
+        return ret;
+        // SourbyCraft end - fix elastic leash behaviour on unleash cancel
     }
 
     @Override
