From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Tue, 20 May 2026 00:00:00 +0700
Subject: [PATCH] SourbyCraft dynamic max stack size


diff --git a/net/minecraft/world/item/Item.java b/net/minecraft/world/item/Item.java
index a97923a94eb8a63b9a2ece6d46b8b19dd4cfb866..f4d840813041436655f0610d982aa60036194aae 100644
--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -108,6 +108,7 @@ public class Item implements FeatureElement, ItemLike {
     public static final Identifier BASE_ATTACK_SPEED_ID = Identifier.withDefaultNamespace("base_attack_speed");
     public static final int DEFAULT_MAX_STACK_SIZE = 64;
     public static final int ABSOLUTE_MAX_STACK_SIZE = 99;
+    public static int sourbycraftMaxStackSize = 99; // SourbyCraft - dynamic max stack
     public static final int MAX_BAR_WIDTH = 13;
     protected static final int APPROXIMATELY_INFINITE_USE_DURATION = 72000;
     private final Holder.Reference<Item> builtInRegistryHolder = BuiltInRegistries.ITEM.createIntrusiveHolder(this);
@@ -152,7 +153,11 @@ public class Item implements FeatureElement, ItemLike {
     }
 
     public int getDefaultMaxStackSize() {
-        return this.components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1);
+        int max = this.components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1);
+        if (max == DEFAULT_MAX_STACK_SIZE && sourbycraftMaxStackSize != DEFAULT_MAX_STACK_SIZE) {
+            return sourbycraftMaxStackSize;
+        }
+        return max;
     }
 
     public void onUseTick(Level level, LivingEntity livingEntity, ItemStack stack, int remainingUseDuration) {
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 76f4bb4368a038dc9660d8af636919cd2095d8d9..e6856f27a97380ecf53b62262fbf1a5a1d1c3118 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -598,7 +598,13 @@ public final class ItemStack implements DataComponentHolder {
     }
 
     public int getMaxStackSize() {
-        return this.getOrDefault(DataComponents.MAX_STACK_SIZE, 1);
+        int max = this.getOrDefault(DataComponents.MAX_STACK_SIZE, 1);
+        // SourbyCraft start - dynamic max stack
+        if (max == Item.DEFAULT_MAX_STACK_SIZE && Item.sourbycraftMaxStackSize != Item.DEFAULT_MAX_STACK_SIZE) {
+            return Item.sourbycraftMaxStackSize;
+        }
+        // SourbyCraft end
+        return max;
     }
 
     public boolean isStackable() {
