From 7daf9bb84557d33068ec36853942aa9fe33a4720 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sat, 6 Jun 2026 17:44:33 +0700
Subject: [PATCH 34/35] SourbyCraft: perf-engine P2 disable saving snowballs

---
 .../projectile/throwableitemprojectile/Snowball.java   | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index 6e82881..247bf9e 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -66,4 +66,14 @@ public class Snowball extends ThrowableItemProjectile {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
+
+    // SourbyCraft start - perf-engine P2 lag-machine: skip NBT save for snowballs
+    @Override
+    public boolean shouldBeSaved() {
+        if (dev.iyanz.sourbycraft.perf.knob.Knobs.LAG_MACHINE_DISABLE_SAVING_SNOWBALLS.get()) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // SourbyCraft end - perf-engine P2 lag-machine
 }
-- 
2.50.1 (Apple Git-155)

