From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <superwhisz@gmail.com>
Date: Sat, 20 Jun 2026 12:53:22 +0700
Subject: [PATCH] SourbyCraft drop final on ServerLevel entityDataController +
 poiDataController for SWM override


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 6e9f91b1fea99985407b61159c1a5273783bc7ac..4a3d2b83a2277840ea96905be5861bdffc9c85b4 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -311,8 +311,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     // Paper start - rewrite chunk system
     private final ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder viewDistanceHolder = new ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder();
     private final ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader chunkLoader = new ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader((ServerLevel)(Object)this);
-    private final ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.EntityDataController entityDataController;
-    private final ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.PoiDataController poiDataController;
+    protected ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.EntityDataController entityDataController; // SourbyCraft - SWM subclass overrides
+    protected ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.PoiDataController poiDataController; // SourbyCraft - SWM subclass overrides
     private final ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.ChunkDataController chunkDataController;
     private final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler;
     private long lastMidTickFailure;
