--- a/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java
+++ b/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java
@@ -86,34 +_,34 @@
 
     @Override
     public String getTitle() {
-        return CraftChatMessage.fromComponent(this.handle.name);
+        return CraftChatMessage.fromComponent(this.handle.getName()); // Canvas - region threading
     }
 
     @Override
     public void setTitle(String title) {
-        this.handle.name = CraftChatMessage.fromString(title, true)[0];
+        this.handle.setName(CraftChatMessage.fromString(title, true)[0]); // Canvas - region threading
         this.handle.broadcast(ClientboundBossEventPacket::createUpdateNamePacket);
     }
 
     @Override
     public BarColor getColor() {
-        return this.convertColor(this.handle.color);
+        return this.convertColor(this.handle.getColor()); // Canvas - region threading
     }
 
     @Override
     public void setColor(BarColor color) {
-        this.handle.color = this.convertColor(color);
+        this.handle.setColor(this.convertColor(color)); // Canvas - region threading
         this.handle.broadcast(ClientboundBossEventPacket::createUpdateStylePacket);
     }
 
     @Override
     public BarStyle getStyle() {
-        return this.convertStyle(this.handle.overlay);
+        return this.convertStyle(this.handle.getOverlay()); // Canvas - region threading
     }
 
     @Override
     public void setStyle(BarStyle style) {
-        this.handle.overlay = this.convertStyle(style);
+        this.handle.setOverlay(this.convertStyle(style)); // Canvas - region threading
         this.handle.broadcast(ClientboundBossEventPacket::createUpdateStylePacket);
     }
 
