--- a/net/minecraft/server/commands/RideCommand.java
+++ b/net/minecraft/server/commands/RideCommand.java
@@ -52,6 +_,24 @@
     }
 
     private static int mount(final CommandSourceStack source, final Entity target, final Entity vehicle) throws CommandSyntaxException {
+    // Canvas start - region threading
+        dev.iyanz.aurora.engine.threadedregions.commands.AbstractCommandExecution._void()
+            .with(target)
+            .executes((Entity entityTarget) -> {
+                Entity entityVehicle = vehicle.getBukkitEntity().getHandleRaw();
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(vehicle)) {
+                    throw dev.iyanz.aurora.engine.threadedregions.commands.AbstractCommandExecution.CANNOT_DO_X_CROSS_REGION.create("mount entities");
+                }
+                mountAction(source, entityTarget, entityVehicle);
+                return (_) -> null;
+            })
+            .noCompletion()
+            .start(source);
+        return Command.SINGLE_SUCCESS;
+    }
+
+    private static int mountAction(final CommandSourceStack source, final Entity target, final Entity vehicle) throws CommandSyntaxException {
+    // Canvas end - region threading
         Entity currentVehicle = target.getVehicle();
         if (currentVehicle != null) {
             throw ERROR_ALREADY_RIDING.create(target.getDisplayName(), currentVehicle.getDisplayName());
@@ -78,13 +_,22 @@
     }
 
     private static int dismount(final CommandSourceStack source, final Entity target) throws CommandSyntaxException {
-        Entity vehicle = target.getVehicle();
-        if (vehicle == null) {
-            throw ERROR_NOT_RIDING.create(target.getDisplayName());
-        }
+    // Canvas start - region threading
+        dev.iyanz.aurora.engine.threadedregions.commands.AbstractCommandExecution._void()
+            .with(target)
+            .executes((Entity entityTarget) -> {
+                Entity vehicle = entityTarget.getVehicle();
+                if (vehicle == null) {
+                    throw ERROR_NOT_RIDING.create(entityTarget.getDisplayName());
+                }
 
-        target.stopRiding();
-        source.sendSuccess(() -> Component.translatable("commands.ride.dismount.success", target.getDisplayName(), vehicle.getDisplayName()), true);
+                entityTarget.stopRiding();
+                source.sendSuccess(() -> Component.translatable("commands.ride.dismount.success", target.getDisplayName(), vehicle.getDisplayName()), true);
+                return (_) -> null;
+            })
+            .noCompletion()
+            .start(source);
+    // Canvas end - region threading
         return Command.SINGLE_SUCCESS;
     }
 }
