--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -150,18 +_,7 @@
 
     private static int teleportToEntity(final CommandSourceStack source, final Collection<? extends Entity> entities, final Entity destination) throws CommandSyntaxException {
         for (Entity entity : entities) {
-            performTeleport(
-                source,
-                entity,
-                (ServerLevel)destination.level(),
-                destination.getX(),
-                destination.getY(),
-                destination.getZ(),
-                EnumSet.noneOf(Relative.class),
-                destination.getYRot(),
-                destination.getXRot(),
-                null
-            );
+            io.papermc.paper.threadedregions.TeleportUtils.teleport(entity, false, destination, Float.valueOf(destination.getYRot()), Float.valueOf(destination.getXRot()), Entity.TELEPORT_FLAG_LOAD_CHUNK, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND, null); // Folia - region threading
         }
 
         if (entities.size() == 1) {
@@ -252,10 +_,29 @@
         double relativeOrAbsoluteX = relatives.contains(Relative.X) ? x - victim.getX() : x;
         double relativeOrAbsoluteY = relatives.contains(Relative.Y) ? y - victim.getY() : y;
         double relativeOrAbsoluteZ = relatives.contains(Relative.Z) ? z - victim.getZ() : z;
-        float relativeOrAbsoluteYRot = relatives.contains(Relative.Y_ROT) ? yRot - victim.getYRot() : yRot;
-        float relativeOrAbsoluteXRot = relatives.contains(Relative.X_ROT) ? xRot - victim.getXRot() : xRot;
+        // Canvas start - region threading
+        float relativeOrAbsoluteYRot = relatives.contains(Relative.Y_ROT) && source.getEntity() != victim ? yRot - victim.getYRot() : yRot;
+        float relativeOrAbsoluteXRot = relatives.contains(Relative.X_ROT) && source.getEntity() != victim ? xRot - victim.getXRot() : xRot;
+        // Canvas end - region threading
         float newYRot = Mth.wrapDegrees(relativeOrAbsoluteYRot);
         float newXRot = Mth.wrapDegrees(relativeOrAbsoluteXRot);
+        // Folia start - region threading
+        if (true) {
+            Vec3 posFinal = new Vec3(x, y, z);
+            Float yawFinal = Float.valueOf(newYRot);
+            Float pitchFinal = Float.valueOf(newXRot);
+            victim.getBukkitEntity().taskScheduler.scheduleOrExecute((Entity nmsEntity) -> { // Canvas - region threading
+                nmsEntity.stopRiding();
+                nmsEntity.teleportAsync(
+                        level, posFinal, yawFinal, pitchFinal, Vec3.ZERO,
+                        org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND,
+                        Entity.TELEPORT_FLAG_LOAD_CHUNK | Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS,
+                        null
+                );
+                }); // Canvas - region threading
+            return;
+        }
+        // Folia end - region threading
         if (victim.teleportTo(level, relativeOrAbsoluteX, relativeOrAbsoluteY, relativeOrAbsoluteZ, relatives, newYRot, newXRot, true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND)) { // Paper - teleport cause
             if (lookAt != null) {
                 lookAt.perform(source, victim);
