--- a/src/main/java/com/destroystokyo/paper/Metrics.java
+++ b/src/main/java/com/destroystokyo/paper/Metrics.java
@@ -593,7 +_,7 @@
             boolean logFailedRequests = config.getBoolean("logFailedRequests", false);
             // Only start Metrics, if it's enabled in the config
             if (config.getBoolean("enabled", true)) {
-                Metrics metrics = new Metrics("Paper", serverUUID, logFailedRequests, Bukkit.getLogger());
+                Metrics metrics = new Metrics("Canvas", serverUUID, logFailedRequests, Bukkit.getLogger()); // Canvas - Rebrand
 
                 metrics.addCustomChart(new Metrics.SimplePie("minecraft_version", () -> {
                     String minecraftVersion = Bukkit.getVersion();
@@ -606,12 +_,23 @@
                 final String paperVersion;
                 final String implVersion = org.bukkit.craftbukkit.Main.class.getPackage().getImplementationVersion();
                 if (implVersion != null) {
-                    final String buildOrHash = implVersion.substring(implVersion.lastIndexOf('-') + 1);
-                    paperVersion = "git-Paper-%s-%s".formatted(Bukkit.getServer().getMinecraftVersion(), buildOrHash);
+                    // Canvas start - Rebrand - new versioning schem for bstats
+                    io.papermc.paper.ServerBuildInfo buildInfo = io.papermc.paper.ServerBuildInfo.buildInfo();
+                    final int num = buildInfo.buildNumber().orElse(-1);
+                    paperVersion = "Canvas-V3-" +
+                        buildInfo.minecraftVersionName() + "-" +
+                        (num != -1 ? num : "DEV");
+                    // Canvas end - Rebrand - new versioning schem for bstats
                 } else {
                     paperVersion = "unknown";
                 }
-                metrics.addCustomChart(new Metrics.SimplePie("paper_version", () -> paperVersion));
+                // Canvas start - Rebrand
+                metrics.addCustomChart(new Metrics.SimplePie("canvas_version", () -> paperVersion));
+                final io.papermc.paper.configuration.GlobalConfiguration paperGlobal = io.papermc.paper.configuration.GlobalConfiguration.get();
+                if (paperGlobal != null) {
+                    metrics.addCustomChart(new Metrics.SimplePie("scheduler_type", () -> paperGlobal.threadedRegions.scheduler.name()));
+                }
+                // Canvas end - Rebrand
 
                 metrics.addCustomChart(new Metrics.DrilldownPie("java_version", () -> {
                     Map<String, Map<String, Integer>> map = new HashMap<>();
