--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
@@ -24,7 +_,7 @@
     // ticket does not expire if the chunk fails to generate before the timeout expires. Rather than blindly adjusting the entire system behavior
     // to fix this small issue, we instead add non-expirable tickets here to keep ender pearls ticking. This is how the original feature should have
     // been implemented, but I don't think Vanilla has proper entity add/remove hooks like we do. Fixes MC-297591
-    private static final TicketType ENDER_PEARL_TICKER = ChunkSystemTicketType.create("chunk_system:ender_pearl_ticker", null, 0L, TicketType.FLAG_LOADING | TicketType.FLAG_SIMULATION | TicketType.FLAG_KEEP_DIMENSION_ACTIVE);
+    private static final TicketType ENDER_PEARL_TICKER = ChunkSystemTicketType.create("chunk_system:ender_pearl_ticker", Integer::compareTo, 0L, TicketType.FLAG_LOADING | TicketType.FLAG_SIMULATION | TicketType.FLAG_KEEP_DIMENSION_ACTIVE); // Folia - region threading
     private final Long2IntOpenHashMap enderPearlChunkCount = new Long2IntOpenHashMap();
     private final boolean keepEnderPearlsTicking;
 
@@ -74,8 +_,10 @@
             ((ChunkSystemServerLevel)this.serverWorld).moonrise$getNearbyPlayers().tickPlayer(player);
         }
         if (entity instanceof ThrownEnderpearl enderpearl && (oldSectionX != newSectionX || oldSectionZ != newSectionZ)) {
-            this.removeEnderPearl(CoordinateUtils.getChunkKey(oldSectionX, oldSectionZ));
-            this.addEnderPearl(CoordinateUtils.getChunkKey(newSectionX, newSectionZ));
+            // Folia start - region threading
+            this.removeEnderPearl(CoordinateUtils.getChunkKey(oldSectionX, oldSectionZ), enderpearl.getId());
+            this.addEnderPearl(CoordinateUtils.getChunkKey(newSectionX, newSectionZ), enderpearl.getId());
+            // Folia end - region threading
         }
         PlatformHooks.get().entityMove(
             entity,
@@ -86,11 +_,12 @@
 
     @Override
     protected void addEntityCallback(final Entity entity) {
+        this.world.getCurrentWorldData().addEntity(entity); // Folia - region threading
         if (entity instanceof ServerPlayer player) {
             ((ChunkSystemServerLevel)this.serverWorld).moonrise$getNearbyPlayers().addPlayer(player);
         }
         if (entity instanceof ThrownEnderpearl enderpearl) {
-            this.addEnderPearl(CoordinateUtils.getChunkKey(enderpearl.chunkPosition()));
+            this.addEnderPearl(CoordinateUtils.getChunkKey(enderpearl.chunkPosition()), enderpearl.getId()); // Folia - region threading
         }
         entity.registerScheduler(); // Paper - optimise Folia entity scheduler
     }
@@ -101,8 +_,9 @@
             ((ChunkSystemServerLevel)this.serverWorld).moonrise$getNearbyPlayers().removePlayer(player);
         }
         if (entity instanceof ThrownEnderpearl enderpearl) {
-            this.removeEnderPearl(CoordinateUtils.getChunkKey(enderpearl.chunkPosition()));
+            this.removeEnderPearl(CoordinateUtils.getChunkKey(enderpearl.chunkPosition()), enderpearl.getId()); // Folia - region threading
         }
+        entity.getBukkitEntity().taskScheduler.registerTo(null); // Folia - region threading
     }
 
     @Override
@@ -130,30 +_,22 @@
         return PlatformHooks.get().screenEntity(this.serverWorld, entity, fromDisk, event);
     }
 
-    private void addEnderPearl(final long coordinate) {
+    private void addEnderPearl(final long coordinate, final int entityId) { // Folia - region threading
         if (!this.keepEnderPearlsTicking) {
             return;
         }
-
-        final int oldCount = this.enderPearlChunkCount.addTo(coordinate, 1);
-        if (oldCount != 0) {
-            return;
-        }
+        // Folia - region threading
         ((ChunkSystemServerLevel)this.serverWorld).moonrise$getChunkTaskScheduler().chunkHolderManager
-            .addTicketAtLevel(ENDER_PEARL_TICKER, coordinate, ChunkHolderManager.ENTITY_TICKING_TICKET_LEVEL, null);
+            .addTicketAtLevel(ENDER_PEARL_TICKER, coordinate, ChunkHolderManager.ENTITY_TICKING_TICKET_LEVEL, Integer.valueOf(entityId)); // Folia - region threading
     }
 
-    private void removeEnderPearl(final long coordinate) {
+    private void removeEnderPearl(final long coordinate, final int entityId) { // Folia - region threading
         if (!this.keepEnderPearlsTicking) {
             return;
         }
 
-        final int oldCount = this.enderPearlChunkCount.addTo(coordinate, -1);
-        if (oldCount != 1) {
-            return;
-        }
-        this.enderPearlChunkCount.remove(coordinate);
+        // Folia - region threading
         ((ChunkSystemServerLevel)this.serverWorld).moonrise$getChunkTaskScheduler().chunkHolderManager
-            .removeTicketAtLevel(ENDER_PEARL_TICKER, coordinate, ChunkHolderManager.ENTITY_TICKING_TICKET_LEVEL, null);
+            .removeTicketAtLevel(ENDER_PEARL_TICKER, coordinate, ChunkHolderManager.ENTITY_TICKING_TICKET_LEVEL, Integer.valueOf(entityId)); // Folia - region threading
     }
 }
