--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -396,7 +_,7 @@
         }
 
         // Folia start - block plugins not marked as supported
-        if (!descriptionFile.isFoliaSupported()) {
+        if (!descriptionFile.isFoliaSupported() && !dev.iyanz.sourbycraft.bridge.AuroraBridge.admitLegacy(descriptionFile.getName())) { // SourbyCraft - Aurora Bridge
             throw new InvalidPluginException("Plugin " + descriptionFile.getFullName() + " is not marked as supporting regionized multithreading");
         }
 
