--- a/canvas-api/build.gradle.kts
+++ b/canvas-api/build.gradle.kts
@@ -121,17 +_,21 @@
         java {
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../canvas-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
+            srcDir(file("../canvas-api/src/main/resources"))
         }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
+            srcDir(file("../canvas-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
+            srcDir(file("../canvas-api/src/test/resources"))
         }
     }
 }
