--- a/net/minecraft/advancements/triggers/BeeNestDestroyedTrigger.java
+++ b/net/minecraft/advancements/triggers/BeeNestDestroyedTrigger.java
@@ -21,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final BlockState state, final ItemStack itemStack, final int numBeesInside) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         this.trigger(player, t -> t.matches(state, itemStack, numBeesInside));
     }
 
