--- a/net/minecraft/advancements/triggers/BrewedPotionTrigger.java
+++ b/net/minecraft/advancements/triggers/BrewedPotionTrigger.java
@@ -16,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final Holder<Potion> potion) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         this.trigger(player, t -> t.matches(potion));
     }
 
