--- a/net/minecraft/advancements/triggers/ConstructBeaconTrigger.java
+++ b/net/minecraft/advancements/triggers/ConstructBeaconTrigger.java
@@ -15,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final int levels) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         this.trigger(player, t -> t.matches(levels));
     }
 
