--- a/net/minecraft/advancements/triggers/ItemDurabilityTrigger.java
+++ b/net/minecraft/advancements/triggers/ItemDurabilityTrigger.java
@@ -17,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final ItemStack itemStack, final int newDurability) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         this.trigger(player, t -> t.matches(itemStack, newDurability));
     }
 
