--- a/net/minecraft/advancements/triggers/LootTableTrigger.java
+++ b/net/minecraft/advancements/triggers/LootTableTrigger.java
@@ -16,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final ResourceKey<LootTable> lootTable) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         this.trigger(player, t -> t.matches(lootTable));
     }
 
