--- a/net/minecraft/advancements/triggers/PlayerInteractTrigger.java
+++ b/net/minecraft/advancements/triggers/PlayerInteractTrigger.java
@@ -20,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final ItemStack itemStack, final Entity interactedWith) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         LootContext context = EntityPredicate.createContext(player, interactedWith);
         this.trigger(player, t -> t.matches(itemStack, context));
     }
