--- a/net/minecraft/advancements/triggers/SpearMobsTrigger.java
+++ b/net/minecraft/advancements/triggers/SpearMobsTrigger.java
@@ -15,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final int number) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         this.trigger(player, t -> t.matches(number));
     }
 
