--- a/net/minecraft/advancements/triggers/SummonedEntityTrigger.java
+++ b/net/minecraft/advancements/triggers/SummonedEntityTrigger.java
@@ -18,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final Entity entity) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         LootContext context = EntityPredicate.createContext(player, entity);
         this.trigger(player, t -> t.matches(context));
     }
