--- a/net/minecraft/advancements/triggers/UsedTotemTrigger.java
+++ b/net/minecraft/advancements/triggers/UsedTotemTrigger.java
@@ -19,6 +_,7 @@
     }
 
     public void trigger(final ServerPlayer player, final ItemStack itemStack) {
+        if (player.level().canvasConfig().disableCriterionTrigger) return; // Canvas - disable criterion trigger config
         this.trigger(player, t -> t.matches(itemStack));
     }
 
