From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: iyanz <boooky10@gmail.com>
Date: Sat, 1 Mar 2025 01:45:59 +0100
Subject: [PATCH] Setup paperweight fork changes


diff --git a/build.gradle.kts b/build.gradle.kts
index 61f9ec2cafd2946bca7edd6ee6f3d1faf7e1b274..fc5522488c4c532295e5f166fdfc31d4ea80e9c0 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -22,6 +22,25 @@ paperweight {
     minecraftVersion = providers.gradleProperty("mcVersion")
     gitFilePatches = false
 
+    // SourbyCraft start - setup paperweight fork
+    val sourbycraft = forks.register("sourbycraft") {
+        serverPatchesDir = rootDirectory.dir("patches/minecraft")
+        featurePatchDir = serverPatchesDir.dir(".")
+        upstream.patchDir("paperServer") {
+            upstreamPath = "paper-server"
+            excludes = setOf("src/minecraft", "patches", "build.gradle.kts")
+            patchesDir = rootDirectory.dir("patches/server")
+            featurePatchDir = patchesDir.dir(".")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+    }
+    activeFork = sourbycraft
+
+    paper {
+        paperServerDir = upstreamsDirectory().map { it.dir("paper/paper-server") } // SourbyCraft - setup paperweight fork
+    }
+    // SourbyCraft end - setup paperweight fork
+
     updatingMinecraft {
         // oldPaperCommit = "d4fe85375af18bfa88f44d7c1e6a61904ae550cc"
     }
@@ -91,12 +110,30 @@ sourceSets {
     main {
         java {
             srcDir("src/main/pufferfish")
+            // SourbyCraft start - configure actual server sources
+            srcDir("../paper-server/src/main/java")
+        }
+        resources {
+            srcDir("../paper-server/src/main/resources")
+        }
+    }
+    test {
+        java {
+            srcDir("../paper-server/src/test/java")
         }
+        resources {
+            srcDir("../paper-server/src/test/resources")
+        }
+        // SourbyCraft end - configure actual server sources
     }
 }
 // Pufferfish end
 
-val log4jPlugins = sourceSets.create("log4jPlugins")
+// SourbyCraft start - configure actual server sources
+val log4jPlugins = sourceSets.create("log4jPlugins") {
+    java { srcDir("../paper-server/src/log4jPlugins/java") }
+}
+// SourbyCraft end - configure actual server sources
 configurations.named(log4jPlugins.compileClasspathConfigurationName) {
     extendsFrom(configurations.compileClasspath.get())
 }
@@ -119,8 +156,8 @@ abstract class MockitoAgentProvider : CommandLineArgumentProvider {
 }
 
 dependencies {
-    implementation(project(":paper-api"))
-    implementation("ca.spottedleaf:leafpile:1.0.0")
+    implementation(projects.sourbycraftApi) // SourbyCraft - update to own api
+    implementation("ca.spottedleaf:leafpile:1.0.0") // SourbyCraft - Paper 26.2 replaced concurrentutil with leafpile
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
@@ -256,7 +293,7 @@ tasks.test {
     jvmArgumentProviders.add(provider)
 }
 
-val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
+val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("../paper-server/src/generated/java").asFile.toPath()
 idea {
     module {
         generatedSourceDirs.add(generatedDir.toFile())
@@ -350,3 +387,12 @@ fill {
         }
     }
 }
+
+// SourbyCraft start - work around strange paperweight madness
+afterEvaluate {
+    if (tasks.names.contains("importSourbycraftLibraryFiles")) {
+        tasks["importSourbycraftLibraryFiles"].mustRunAfter(tasks["rebuildMinecraftResourcePatches"])
+        tasks["collectSourbycraftATsFromPatches"].mustRunAfter(tasks["rebuildMinecraftResourcePatches"])
+    }
+}
+// SourbyCraft end - work around strange paperweight madness
