From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Wed, 13 May 2026 13:53:29 +0700
Subject: [PATCH] fix: Specification-Version uses implementationVersion


diff --git a/build.gradle.kts b/build.gradle.kts
index 0e32a7f6edb465310583ec0704fb8fa7d6886101..81cf27833b78b210f517d5ca4921c4504403d21f 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -268,7 +268,7 @@ tasks.jar {
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
             "Specification-Title" to "SourbyCraft", // SourbyCraft - branding
-            "Specification-Version" to project.version,
+            "Specification-Version" to implementationVersion,
             "Specification-Vendor" to "SourbyCraft", // SourbyCraft - branding
             "Brand-Id" to "sourbycraft:sourbycraft", // SourbyCraft - branding
             "Brand-Name" to "SourbyCraft", // SourbyCraft - branding
