From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Wed, 13 May 2026 18:43:02 +0700
Subject: [PATCH] fix: escape dot in split regex


diff --git a/build.gradle.kts b/build.gradle.kts
index db6fbf56f2d57e48429889ee0403aca9bf788d97..b56c68c50b8bd13551eef08c0c4c232694f3bd43 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -224,7 +224,7 @@ tasks.jar {
         val mcVersion = rootProject.providers.gradleProperty("mcVersion").get()
         val build = System.getenv("BUILD_NUMBER") ?: null
         val releaseVersionFull = rootProject.providers.gradleProperty("releaseVersion").get()
-        val releaseMajor = releaseVersionFull.split(".")[0]
+        val releaseMajor = releaseVersionFull.split("\\.")[0]
         val codename = rootProject.providers.gradleProperty("codename").getOrElse("dev")
         val now = ZonedDateTime.now(ZoneId.of("Asia/Jakarta"))
         val buildTimeIso = now.toInstant().toString()
