From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Fri, 15 May 2026 01:30:00 +0700
Subject: [PATCH] feat: version suffix from codename


diff --git a/build.gradle.kts b/build.gradle.kts
index 6f6575ec9ccb96834a9471446a473b29edae10ee..e32616ec21895e9deb1e45bbb8ac6b3d411441ea 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -260,6 +260,7 @@ tasks.jar {
         val suffix = when {
             gitBranch.contains("experimental") || gitBranch.contains("feat") -> "EXP"
             gitBranch.contains("-dev") || gitBranch.contains("develop") -> "DEV"
+            codename == "dev" -> "DEV"
             else -> "REL"
         }
         val implementationVersion = when {
