From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Fri, 15 May 2026 01:30:00 +0700
Subject: [PATCH] feat: version suffix from codename


diff --git a/build.gradle.kts b/build.gradle.kts
index b56c68c50b8bd13551eef08c0c4c232694f3bd43..17d9f794c9730091f5897bcbf41f00b4cb628c13 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -233,6 +233,7 @@ tasks.jar {
         val suffix = when {
             gitBranch.contains("experimental") || gitBranch.contains("feat") -> "EXP"
             gitBranch.contains("-dev") || gitBranch.contains("develop") -> "DEV"
+            codename == "dev" -> "DEV"
             else -> "REL"
         }
         val implementationVersion = when {
