From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Sun, 14 Jun 2026 08:20:12 +0700
Subject: [PATCH] fix: set oldPaperCommit for updatingMinecraft


diff --git a/build.gradle.kts b/build.gradle.kts
index 6fd8c84095af5dd2857a8ee080074ee8cded4971..1c8ed21e77a4feb7a74d26bbbb7b6c2ea1446440 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -45,7 +45,7 @@ paperweight {
     // SourbyCraft end - setup paperweight fork
 
     updatingMinecraft {
-        // oldPaperCommit = "711c5de2b05df39b72c44ff54e9f9381f8d153cb"
+        oldPaperCommit = "bd74bf6581ce81e59bdab07eadbfbe5d485eefa7"
     }
 }
 
