From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin Raneri <kevin.raneri@gmail.com>
Date: Tue, 9 Nov 2021 23:36:56 -0500
Subject: [PATCH] Pufferfish API Buildscript Patches


diff --git a/build.gradle.kts b/build.gradle.kts
index 49b24787796658987f2b04c433d0316e7a23bb67..b7b9a2d4667d62fc186c4efb3041f1f9c496f305 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -89,6 +89,8 @@ dependencies {
     testImplementation("org.ow2.asm:asm-tree:9.8")
     mockitoAgent("org.mockito:mockito-core:5.14.1") { isTransitive = false } // configure mockito agent that is needed in newer java versions
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
+
+    api("io.sentry:sentry:7.15.0") // Pufferfish
 }
 
 val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
@@ -101,6 +103,7 @@ sourceSets {
     main {
         java {
             srcDir(generatedDir)
+            srcDir("src/main/pufferfish") // Pufferfish
         }
     }
 }
@@ -166,6 +169,12 @@ val generateApiVersioningFile = tasks.register<GenerateApiVersioningFile>("gener
     apiVersion.set(rootProject.providers.gradleProperty("apiVersion"))
 }
 
+// Pufferfish start
+tasks.withType<JavaCompile> {
+    options.compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish end
+
 tasks.jar {
     from(generateApiVersioningFile.flatMap { it.outputFile })
     manifest {
@@ -183,6 +192,7 @@ val services = objects.newInstance<Services>()
 
 tasks.withType<Javadoc>().configureEach {
     val options = options as StandardJavadocDocletOptions
+    options.addStringOption("-add-modules", "jdk.incubator.vector") // Pufferfish
     options.overview = "src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
