--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -94,11 +_,11 @@
             Optional<BlockPos> pos = leashData.delayedLeashInfo.right();
             if (leashUuid.isPresent()) {
                 Entity leasher = serverLevel.getEntity(leashUuid.get());
-                if (leasher != null) {
+                if (leasher != null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(leasher)) { // Canvas - region threading
                     setLeashedTo(entity, leasher, true);
                     return;
                 }
-            } else if (pos.isPresent()) {
+            } else if (pos.isPresent() && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity.level(), pos.get())) { // Canvas - region threading
                 setLeashedTo(entity, LeashFenceKnotEntity.getOrCreateKnot(serverLevel, pos.get()), true);
                 return;
             }
