From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Sat, 18 Apr 2026 10:49:40 +0800
Subject: [PATCH] Add tick regions api


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 8811e9ccb113ebb1f7e66814f363e0747b58d7ef..5fc4b44c05b34b64dfdfc1fe996124ec03490465 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2077,4 +2077,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    // Luminol start - Tick regions api
+    @Override
+    public me.earthme.luminol.api.ThreadedRegionizer getThreadedRegionizer() {
+        return this.world.regioniser.threadedRegionizerAPI;
+    }
+    // Luminol end
 }
