--- a/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -55,7 +_,7 @@
     public void tick() {
         this.time++;
         this.applyEffectsFromBlocks();
-        this.handlePortal();
+        // this.handlePortal(); // Folia - region threading
         if (this.level() instanceof ServerLevel) {
             BlockPos pos = this.blockPosition();
             if (((ServerLevel)this.level()).getDragonFight() != null && this.level().getBlockState(pos).isAir()) {
