--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -11,8 +_,6 @@
 import net.minecraft.util.Mth;
 import net.minecraft.util.debug.DebugSubscriptions;
 import net.minecraft.util.debug.ServerDebugSubscribers;
-import net.minecraft.util.profiling.Profiler;
-import net.minecraft.util.profiling.ProfilerFiller;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.ai.attributes.Attributes;
@@ -99,12 +_,12 @@
     }
 
     public void recomputePath() {
-        if (this.level.getGameTime() - this.timeLastRecompute <= 20L || !this.canUpdatePath()) {
+        if (this.tick - this.timeLastRecompute <= 20L || !this.canUpdatePath()) { // Folia - region threading
             this.hasDelayedRecomputation = true;
         } else if (this.targetPos != null) {
             this.path = null;
             this.path = this.createPath(this.targetPos, this.reachRange);
-            this.timeLastRecompute = this.level.getGameTime();
+            this.timeLastRecompute = this.tick; // Folia - region threading
             this.hasDelayedRecomputation = false;
         }
     }
@@ -174,8 +_,15 @@
         // Paper start - EntityPathfindEvent
         boolean copiedSet = false;
         for (BlockPos possibleTarget : targets) {
-            if (!this.mob.level().getWorldBorder().isWithinBounds(possibleTarget) || !new com.destroystokyo.paper.event.entity.EntityPathfindEvent(this.mob.getBukkitEntity(), // Paper - don't path out of world border
-                org.bukkit.craftbukkit.util.CraftLocation.toBukkit(possibleTarget, this.mob.level()), target == null ? null : target.getBukkitEntity()).callEvent()) {
+            // Canvas start - region threading
+            // expands the if statement too for readability
+            if (
+                !this.mob.level().getWorldBorder().isWithinBounds(possibleTarget) || // Paper - don't path out of world border
+                // thread check before event call as to not pass out of region targets to the event
+                !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.mob.level(), possibleTarget) ||
+                !new com.destroystokyo.paper.event.entity.EntityPathfindEvent(this.mob.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(possibleTarget, this.mob.level()), target == null ? null : target.getBukkitEntity()).callEvent()
+            ) {
+            // Canvas end - region threading
                 if (!copiedSet) {
                     copiedSet = true;
                     targets = new java.util.HashSet<>(targets);
@@ -188,13 +_,10 @@
             }
         }
         // Paper end - EntityPathfindEvent
-        ProfilerFiller profiler = Profiler.get();
-        profiler.push("pathfind");
         BlockPos fromPos = above ? this.mob.blockPosition().above() : this.mob.blockPosition();
         int radius = (int)(maxPathLength + radiusOffset);
         PathNavigationRegion region = new PathNavigationRegion(this.level, fromPos.offset(-radius, -radius, -radius), fromPos.offset(radius, radius, radius));
         Path path = this.pathFinder.findPath(region, this.mob, targets, maxPathLength, reachRange, this.maxVisitedNodesMultiplier);
-        profiler.pop();
         if (path != null && path.getTarget() != null) {
             this.targetPos = path.getTarget();
             this.reachRange = reachRange;
@@ -219,7 +_,7 @@
 
     public boolean moveTo(final Entity target, final double speedModifier) {
         // Paper start - Perf: Optimise pathfinding
-        if (this.pathfindFailures > 10 && this.path == null && net.minecraft.server.MinecraftServer.currentTick < this.lastFailure + 40) {
+        if (this.pathfindFailures > 10 && this.path == null && this.tick < this.lastFailure + 40) { // Folia - region threading
             return false;
         }
         // Paper end - Perf: Optimise pathfinding
@@ -231,7 +_,7 @@
             return true;
         } else {
             this.pathfindFailures++;
-            this.lastFailure = net.minecraft.server.MinecraftServer.currentTick;
+            this.lastFailure = this.tick; // Folia - region threading
             return false;
         }
         // Paper end - Perf: Optimise pathfinding
@@ -286,6 +_,20 @@
 
             if (!this.isDone()) {
                 Vec3 target = this.path.getNextEntityPos(this.mob);
+                // Canvas start - region threading
+
+                // the next entity pos can be out of region(stale?), which basically deletes
+                // the mob that is pathfinding whenever we try and get that location when it's
+                // out of region assuming the chunk is unloaded
+
+                // thread check here so that we don't set an out of region position
+                // sorry to the many horses that died on MagnoliaMC bc of this :sob:
+
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.mob.level(), target)) {
+                    this.hasDelayedRecomputation = true; // set this so next tick will recompute path
+                    return;
+                }
+                // Canvas end - region threading
                 this.mob.getMoveControl().setWantedPosition(target.x, this.getGroundY(target), target.z, this.speedModifier);
             }
         }
