From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 28 Jun 2026 20:57:48 +0800
Subject: [PATCH] Leaves: Disable packet limit


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index ecaa296f576c5c90fda89e0406d6552628ef5de3..d2e9d46bbd9ee4ce45419b11499d7a7bc50f0219 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -258,8 +258,8 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             if (this.stopReadingPackets) {
                 return;
             }
-            if (this.allPacketCounts != null ||
-                io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.overrides.containsKey(packet.getClass())) {
+            if (!me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable && (this.allPacketCounts != null || // Luminol - Add config to force disable the packet limiter of Paper
+                io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.overrides.containsKey(packet.getClass()))) { // Luminol - Add config to force disable the packet limiter of Paper
                 long time = System.nanoTime();
                 synchronized (PACKET_LIMIT_LOCK) {
                     if (this.allPacketCounts != null) {
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 022d269b10ca0d34853c6d771943305b04442370..5ffd2da7ccf574dee21bffe57ffe4ad721d4e4a3 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -879,7 +879,7 @@ public class ServerGamePacketListenerImpl
     public void handleCustomCommandSuggestions(final ServerboundCommandSuggestionPacket packet) {
         // PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level()); // Paper - AsyncTabCompleteEvent; run this async
         // CraftBukkit start
-        if (!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.nameAndId()) && !this.server.isSingleplayerOwner(this.player.nameAndId())) { // Paper - configurable tab spam limits
+        if (!me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable && !this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.nameAndId()) && !this.server.isSingleplayerOwner(this.player.nameAndId())) { // Paper - configurable tab spam limits // Leaves - can disable
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - Kick event cause // Paper - add proper async disconnect
             return;
         }
@@ -2115,6 +2115,7 @@ public class ServerGamePacketListenerImpl
     private long lastLimitedPacket = -1;
 
     private boolean checkLimit(long timestamp) {
+        if (me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable) return true; // Leaves - disable
         if (!io.papermc.paper.configuration.GlobalConfiguration.get().spamLimiter.incomingPacketThreshold.enabled()) {
             return true;
         }
@@ -2681,6 +2682,8 @@ public class ServerGamePacketListenerImpl
 
     // Spigot start - spam exclusions
     private void detectRateSpam(final TickThrottler throttler, final String message) {
+        if (me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable) return; // Leaves - disable
+        // CraftBukkit start - replaced with thread safe throttle
         if (org.spigotmc.SpigotConfig.enableSpamExclusions) {
             for (String exclude : org.spigotmc.SpigotConfig.spamExclusions) {
                 if (exclude != null && message.startsWith(exclude)) {
@@ -3479,7 +3482,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handlePlaceRecipe(final ServerboundPlaceRecipePacket packet) {
         // Paper start - auto recipe limit
-        if (!org.bukkit.Bukkit.isPrimaryThread()) {
+        if (!me.earthme.luminol.config.modules.misc.PaperPacketLimiterConfig.forceDisable && !org.bukkit.Bukkit.isPrimaryThread()) { // Leaves - can disable
             if (!this.recipeSpamPackets.isIncrementAndUnderThreshold()) {
                 this.disconnectAsync(net.minecraft.network.chat.Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - kick event cause // Paper - add proper async disconnect
                 return;
