From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sun, 12 Jan 2025 10:18:38 +0800
Subject: [PATCH] Rebrand to Luminol


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..3cb00df3f455c18b3c75b7401a58472888ace7b4 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -19,6 +19,17 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    // SourbyCraft start
+    /**
+     * The brand id for folia
+     */
+    Key BRAND_FOLIA_ID = Key.key("papermc", "folia");
+    /**
+     * The brand id for SourbyCraft.
+     */
+    Key BRAND_SOURBYCRAFT_ID = Key.key("sourbycraft", "sourbycraft");
+    // SourbyCraft end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
