--- a/net/minecraft/commands/arguments/selector/EntitySelector.java
+++ b/net/minecraft/commands/arguments/selector/EntitySelector.java
@@ -184,11 +_,25 @@
     private void addEntities(final List<Entity> result, final ServerLevel level, final @Nullable AABB absoluteAABB, final Predicate<Entity> predicate) {
         int limit = this.getResultLimit();
         if (result.size() < limit) {
-            if (absoluteAABB != null) {
-                level.getEntities(this.type, absoluteAABB, predicate, result, limit);
-            } else {
-                level.getEntities(this.type, predicate, result, limit);
-            }
+            // Canvas start - region threading
+
+            // people can use the params in the entity selector to create a bounding
+            // box large enough that it exceeds the region boundaries, so we modify
+            // the collecting of entities here to avoid the thread check, and we just
+            // check the entity is actually contained in the AABB
+
+            level.getEntities(this.type, (entity) -> {
+                // check AABB first, then check predicate
+                if (absoluteAABB != null && !entity.getBoundingBox().intersects(absoluteAABB)) {
+                    return false;
+                }
+
+                // if the AABB isn't null and intersects, the entity is contained within the box
+                // and should be ok for collection if the predicate passes
+
+                return predicate.test(entity);
+            }, result, limit);
+            // Canvas end - region threading
         }
     }
 
