--- /dev/null
+++ b/io/papermc/paper/threadedregions/FoliaWatchdogThread.java
@@ -1,0 +_,147 @@
+package io.papermc.paper.threadedregions;
+
+import com.mojang.logging.LogUtils;
+import io.canvasmc.canvas.util.Util;
+import io.papermc.paper.ServerBuildInfo;
+import io.papermc.paper.threadedregions.TickRegionScheduler.RegionScheduleHandle;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MonitorInfo;
+import java.lang.management.OperatingSystemMXBean;
+import java.lang.management.RuntimeMXBean;
+import java.lang.management.ThreadInfo;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.bukkit.Bukkit;
+import org.jspecify.annotations.NonNull;
+import org.slf4j.Logger;
+
+public final class FoliaWatchdogThread extends Thread {
+
+    private static final Logger LOGGER = LogUtils.getClassLogger();
+    public static final String BREAK_LINE = "----------------------------------------------------------"; // Canvas - more detailed watchdog output
+
+    public static final class RunningTick {
+
+        public final long start;
+        public final RegionScheduleHandle handle;
+        public final Thread thread;
+
+        private long lastPrint;
+
+        public RunningTick(final long start, final RegionScheduleHandle handle, final Thread thread) {
+            this.start = start;
+            this.handle = handle;
+            this.thread = thread;
+            this.lastPrint = start;
+        }
+    }
+
+    private final LinkedHashSet<RunningTick> ticks = new LinkedHashSet<>();
+
+    public FoliaWatchdogThread() {
+        super("Folia Watchdog Thread");
+        this.setDaemon(true);
+        this.setUncaughtExceptionHandler((final Thread thread, final Throwable throwable) -> {
+            LOGGER.error("Uncaught exception in thread '{}'", thread.getName(), throwable);
+        });
+    }
+
+    @Override
+    public void run() {
+        for (;;) {
+            try {
+                Thread.sleep(1000L);
+            } catch (final InterruptedException _) {}
+
+            if (RegionShutdownThread.isShuttingDown()) {
+                return;
+            }
+
+            if (!TickRegions.hasStarted()) {
+                continue;
+            }
+
+            final List<RunningTick> ticks;
+            synchronized (this.ticks) {
+                if (this.ticks.isEmpty()) {
+                    continue;
+                }
+                ticks = new ArrayList<>(this.ticks);
+            }
+
+            final long now = System.nanoTime();
+
+            for (final RunningTick tick : ticks) {
+                final long elapsed = now - tick.lastPrint;
+                if (elapsed <= TimeUnit.SECONDS.toNanos(5L)) {
+                    continue;
+                }
+                tick.lastPrint = now;
+
+                final String totalElapsedS = Util.formatNanosToLargestDecimalUnit(now - tick.start); // Canvas - region threading
+
+                if (tick.handle instanceof TickRegions.ConcreteRegionTickHandle region) {
+                    LOGGER.error("Tick region located in world '{}' around chunk '{}' has not responded in {}s:",
+                        region.region.world.getWorld().getName(), region.region.region.getCenterChunk(), totalElapsedS
+                    );
+                } else {
+                    // assume global
+                    LOGGER.error("Global region has not responded in {}s:", totalElapsedS);
+                }
+
+                // Canvas start - more detailed watchdog output
+                dumpVersionInfo();
+                dumpThread(ManagementFactory.getThreadMXBean().getThreadInfo(tick.thread.threadId(), Integer.MAX_VALUE));
+                // Canvas end - more detailed watchdog output
+            }
+        }
+    }
+    // Canvas start - more detailed watchdog output
+
+    private static void dumpThread(final @NonNull ThreadInfo threadInfo) {
+        //noinspection LoggingSimilarMessage
+        LOGGER.error(BREAK_LINE);
+        LOGGER.error("Current Thread: {}", threadInfo.getThreadName());
+        LOGGER.error("\tPID: {} | Suspended: {} | Native: {} | State: {}",
+            threadInfo.getThreadId(),
+            threadInfo.isSuspended(),
+            threadInfo.isInNative(),
+            threadInfo.getThreadState()
+        );
+
+        final MonitorInfo[] lockedMonitors = threadInfo.getLockedMonitors();
+        if (lockedMonitors.length > 0) {
+            LOGGER.error("\tThread is waiting on {} monitor(s):", lockedMonitors.length);
+            for (final MonitorInfo lockedMonitor : lockedMonitors) {
+                LOGGER.error("\t\tLocked on:{}", lockedMonitor.getLockedStackFrame());
+            }
+        }
+
+        LOGGER.error("\tStack:");
+        for (final StackTraceElement stackElement : threadInfo.getStackTrace()) {
+            LOGGER.error("\t\t{}", stackElement);
+        }
+    }
+
+    private static void dumpVersionInfo() {
+        //noinspection LoggingSimilarMessage
+        LOGGER.error(BREAK_LINE);
+        LOGGER.error("- DO NOT REPORT THIS TO PAPER/FOLIA, REPORT TO {} INSTEAD -", ServerBuildInfo.buildInfo().brandName().toUpperCase());
+        LOGGER.error("Version: {}", Bukkit.getVersion());
+    }
+    // Canvas end - more detailed watchdog output
+
+    public void addTick(final RunningTick tick) {
+        synchronized (this.ticks) {
+            this.ticks.add(tick);
+        }
+    }
+
+    public void removeTick(final RunningTick tick) {
+        synchronized (this.ticks) {
+            this.ticks.remove(tick);
+        }
+    }
+}
