--- a/src/main/java/io/canvasmc/canvas/WorldConfig.java
+++ b/src/main/java/io/canvasmc/canvas/WorldConfig.java
@@ -47,7 +_,7 @@
     // we have a logger internally here for level-config related things, and should not be used globally. the global
     // config class should be the logger publicly used
 
-    private static final Logger LOGGER = LoggerFactory.getLogger("CanvasWorlds");
+    private static final Logger LOGGER = LoggerFactory.getLogger("SourbyCraft"); // SourbyCraft - rebrand console prefix [CanvasWorlds] -> [SourbyCraft]
 
     private static final Path BASE_FILE = Path.of("config/canvas-worlds.yml").toAbsolutePath().normalize();
 
