From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: iyanz <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 20:30:37 +0200
Subject: [PATCH] Add SourbyCraft configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 718861fccd0f7a0ceb550c96d4e9691eb1bb6804..46420d68351632caa7ba6c4a060cba0939d1aad1 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2425,6 +2425,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // SourbyCraft start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSourbyCraftConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // SourbyCraft end
+
         /**
          * Sends the component to the player
          *
