From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: iyanz <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 20:30:37 +0200
Subject: [PATCH] Add SourbyCraft configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index dd8ad5976eb413f049245a4f37bd94b3e41829de..1bcbf0ab7e89db4631c6f4541364433811da97f7 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2444,6 +2444,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // SourbyCraft start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSourbyCraftConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // SourbyCraft end
+
         /**
          * Sends the component to the player
          *
