From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yan <Yan>
Date: Mon, 15 Jun 2026 15:45:29 +0700
Subject: [PATCH] Populate despawn time for existing entities on reload


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 86ae93a9cad3d165bbe0aec5babbe98dbf50d625..d384e9d92a85a8de29ba7910626052e8d1c3a4f0 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -209,6 +209,13 @@ public class WorldConfiguration extends ConfigurationPart {
                     range.hard().preComputed(category.getDespawnDistance(), category.getSerializedName());
                     range.soft().preComputed(category.getNoDespawnDistance(), category.getSerializedName());
                 }
+
+                // SourbyCraft start - populate despawn time for existing entities on reload
+                net.minecraft.server.MinecraftServer.getServer().getAllLevels().forEach(level ->
+                        level.getAllEntities().forEach(entity ->
+                                entity.despawnTime = entity.getType() == EntityType.PLAYER ? -1 :
+                                        this.despawnTime.getOrDefault(entity.getType(), IntOr.Disabled.DISABLED).or(-1)));
+                // SourbyCraft end - populate despawn time for existing entities on reload
             }
 
             public WaterAnimalSpawnHeight wateranimalSpawnHeight;
