From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: iyanz <boooky10@gmail.com>
Date: Fri, 25 Jun 2021 20:37:53 +0200
Subject: [PATCH] Changed branding


diff --git a/src/main/java/com/destroystokyo/paper/Metrics.java b/src/main/java/com/destroystokyo/paper/Metrics.java
index 9f4d73fd321f6f0cd19fdd1b63068d2ea6c0816b..96917a0a72ec3d17c930e8023ee0d62b4bcf661a 100644
--- a/src/main/java/com/destroystokyo/paper/Metrics.java
+++ b/src/main/java/com/destroystokyo/paper/Metrics.java
@@ -593,7 +593,7 @@ public class Metrics {
             boolean logFailedRequests = config.getBoolean("logFailedRequests", false);
             // Only start Metrics, if it's enabled in the config
             if (config.getBoolean("enabled", true)) {
-                Metrics metrics = new Metrics("Paper", serverUUID, logFailedRequests, Bukkit.getLogger());
+                Metrics metrics = new Metrics("SourbyCraft", serverUUID, logFailedRequests, Bukkit.getLogger()); // SourbyCraft - branding
 
                 metrics.addCustomChart(new Metrics.SimplePie("minecraft_version", () -> {
                     String minecraftVersion = Bukkit.getVersion();
@@ -607,11 +607,11 @@ public class Metrics {
                 final String implVersion = org.bukkit.craftbukkit.Main.class.getPackage().getImplementationVersion();
                 if (implVersion != null) {
                     final String buildOrHash = implVersion.substring(implVersion.lastIndexOf('-') + 1);
-                    paperVersion = "git-Paper-%s-%s".formatted(Bukkit.getServer().getMinecraftVersion(), buildOrHash);
+                    paperVersion = "git-SourbyCraft-%s-%s".formatted(Bukkit.getServer().getMinecraftVersion(), buildOrHash); // SourbyCraft - branding
                 } else {
                     paperVersion = "unknown";
                 }
-                metrics.addCustomChart(new Metrics.SimplePie("paper_version", () -> paperVersion));
+                metrics.addCustomChart(new Metrics.SimplePie("sourbycraft_version", () -> paperVersion)); // SourbyCraft - branding
 
                 metrics.addCustomChart(new Metrics.DrilldownPie("java_version", () -> {
                     Map<String, Map<String, Integer>> map = new HashMap<>();
diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index 602d3ce2cd16405f35453ce12a207b70312a4a8a..efd31090fec2e594d61d9618cf61357ec977baa2 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -40,10 +40,10 @@ public class PaperVersionFetcher implements VersionFetcher {
     private static final ComponentLogger COMPONENT_LOGGER = ComponentLogger.logger(LogManager.getRootLogger().getName());
     private static final int DISTANCE_ERROR = -1;
     private static final int DISTANCE_UNKNOWN = -2;
-    private static final String DOWNLOAD_PAGE = "https://papermc.io/downloads/paper";
-    private static final String REPOSITORY = "PaperMC/Paper";
+    private static final String DOWNLOAD_PAGE = "https://github.com/YanIanZ/SourbyCraft/releases"; // SourbyCraft - branding
+    private static final String REPOSITORY = "YanIanZ/SourbyCraft"; // SourbyCraft - branding
     private static final ServerBuildInfo BUILD_INFO = ServerBuildInfo.buildInfo();
-    private static final String USER_AGENT = BUILD_INFO.brandName() + "/" + BUILD_INFO.asString(VERSION_SIMPLE) + " (https://papermc.io)";
+    private static final String USER_AGENT = BUILD_INFO.brandName() + "/" + BUILD_INFO.asString(VERSION_SIMPLE) + " (https://github.com/YanIanZ/SourbyCraft)"; // SourbyCraft - branding
     private static final Gson GSON = new Gson();
 
     @Override
@@ -72,7 +72,7 @@ public class PaperVersionFetcher implements VersionFetcher {
             COMPONENT_LOGGER.warn(text("*** You are running a development version without access to version information ***"));
         } else {
             final Optional<MinecraftVersionFetcher> apiResult = fetchMinecraftVersionList();
-            if (buildNumber.isPresent()) {
+            if (false && buildNumber.isPresent()) { // SourbyCraft - use github for version fetching
                 distance = fetchDistanceFromSiteApi(buildNumber.getAsInt());
             } else {
                 final Optional<String> gitBranch = BUILD_INFO.gitBranch();
diff --git a/src/main/java/com/destroystokyo/paper/console/PaperConsole.java b/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
index 6ee39b534b8d992655bc0cef3c299d12cbae0034..291d45623d7e55f12dddb17e2c665562d0eb0567 100644
--- a/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
+++ b/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
@@ -20,7 +20,7 @@ public final class PaperConsole extends SimpleTerminalConsole {
     @Override
     protected LineReader buildReader(LineReaderBuilder builder) {
         builder
-                .appName("Paper")
+                .appName("SourbyCraft") // SourbyCraft - branding
                 .variable(LineReader.HISTORY_FILE, java.nio.file.Paths.get(".console_history"))
                 .completer(new ConsoleCommandCompleter(this.server))
                 .option(LineReader.Option.COMPLETE_IN_WORD, true);
diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index 74ffdc823e66fc5ec027c4b7c462382bcbfe2be2..347f20a5107eb3644b4e42adfccbd7cc181ea512 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -31,6 +31,7 @@ public record ServerBuildInfoImpl(
     private static final String ATTRIBUTE_GIT_COMMIT = "Git-Commit";
 
     private static final String BRAND_PAPER_NAME = "Paper";
+    private static final String BRAND_SOURBYCRAFT_NAME = "SourbyCraft"; // SourbyCraft - branding
 
     private static final String BUILD_DEV = "DEV";
 
@@ -42,9 +43,9 @@ public record ServerBuildInfoImpl(
         this(
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_ID)
                 .map(Key::key)
-                .orElse(BRAND_PAPER_ID),
+                .orElse(BRAND_SOURBYCRAFT_ID), // SourbyCraft - branding
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_NAME)
-                .orElse(BRAND_PAPER_NAME),
+                .orElse(BRAND_SOURBYCRAFT_NAME), // SourbyCraft - branding
             SharedConstants.getCurrentVersion().id(),
             SharedConstants.getCurrentVersion().name(),
             getManifestAttribute(manifest, ATTRIBUTE_BUILD_NUMBER)
@@ -61,6 +62,7 @@ public record ServerBuildInfoImpl(
 
     @Override
     public boolean isBrandCompatible(final @NotNull Key brandId) {
+        if (brandId.equals(BRAND_PAPER_ID)) return true; // SourbyCraft - branding; mark paper as compatible
         return brandId.equals(this.brandId);
     }
 
