From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 8 Dec 2024 03:12:18 +0100
Subject: [PATCH] Populate despawn time for existing entities on reload


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index d34e627f8b6711d443553af74113d096b783b36d..c1be52267f89bd547550112ab697cbd9809a4fa5 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -208,6 +208,13 @@ public class WorldConfiguration extends ConfigurationPart {
                     range.hard().preComputed(category.getDespawnDistance(), category.getSerializedName());
                     range.soft().preComputed(category.getNoDespawnDistance(), category.getSerializedName());
                 }
+
+                // CloudPlane start - populate despawn time for existing entities on reload
+                net.minecraft.server.MinecraftServer.getServer().getAllLevels().forEach(level ->
+                        level.getAllEntities().forEach(entity ->
+                                entity.despawnTime = entity.getType() == EntityType.PLAYER ? -1 :
+                                        this.despawnTime.getOrDefault(entity.getType(), IntOr.Disabled.DISABLED).or(-1)));
+                // CloudPlane end - populate despawn time for existing entities on reload
             }
 
             public WaterAnimalSpawnHeight wateranimalSpawnHeight;
