--- a/src/main/java/io/papermc/paper/datapack/Datapack.java
+++ b/src/main/java/io/papermc/paper/datapack/Datapack.java
@@ -28,6 +_,7 @@
      * @param enabled true to enable, false to disable
      * @apiNote This method may be deprecated in the future as setters on a "snapshot" type are undesirable.
      */
+    @io.canvasmc.canvas.Unsupported // Canvas - region threading
     void setEnabled(boolean enabled);
 
     /**
