--- a/net/minecraft/core/component/DataComponentExactPredicate.java
+++ b/net/minecraft/core/component/DataComponentExactPredicate.java
@@ -102,7 +_,7 @@
     }
 
     public static class Builder {
-        private final List<TypedDataComponent<?>> expectedComponents = new ArrayList<>();
+        public List<TypedDataComponent<?>> expectedComponents = new ArrayList<>();
 
         Builder() {
         }
