From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Mon, 27 Jan 2025 13:01:59 +0800
Subject: [PATCH] Tick regions api


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index ace05559319506ab1f20dee8b4050cd797ca695b..2631c556924b456348b3673cad6c76fa59de56a8 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4691,4 +4691,8 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
             }
         }
     }
+
+    // Luminol start - Tick regions api
+    me.earthme.luminol.api.ThreadedRegionizer getThreadedRegionizer();
+    // Luminol end - Tick regions api
 }
